--- run.cns	2020-02-20 11:28:46.882095042 +0100
+++ ../run_ref_for_patch.cns	2020-02-20 11:28:29.309997791 +0100
@@ -567,16 +567,16 @@
 
 {==================== histidine patches =====================}
 {* Automatically define histidine protonation state based on energetics *}
-{===>} autohis=true;
+{===>} autohis=false;
 
 {* Patch to change doubly protonated HIS to singly protonated histidine (HD1) *}
 {* just give the residue number of the histidines for the HISD patch, set them to zero if you don't want them *}
 
 {* Number of HISD for molecule 1 *}
-{===>} numhisd_1=0;
+{===>} numhisd_1=2;
 
-{===>} hisd_1_1=0;
-{===>} hisd_1_2=0;
+{===>} hisd_1_1=36;
+{===>} hisd_1_2=109;
 {===>} hisd_1_3=0;
 {===>} hisd_1_4=0;
 {===>} hisd_1_5=0;
@@ -1156,9 +1156,9 @@
 {* just give the residue number of the histidines for the HISD patch, set them to zero if you don't want them *}
 
 {* Number of HISE for molecule 1 *}
-{===>} numhise_1=0;
+{===>} numhise_1=1;
 
-{===>} hise_1_1=0;
+{===>} hise_1_1=113;
 {===>} hise_1_2=0;
 {===>} hise_1_3=0;
 {===>} hise_1_4=0;
@@ -2327,12 +2327,12 @@
 {* Fully flexible segments for molecule 2                            *}
 {* Note that current max is 5 (edit the run.cns to add more segments *}
 
-{===>} nfle_2=0;
+{===>} nfle_2=1;
 
 {* Residues of molecule 2 at interface *}
 {+ table: rows=5 "segment 1" "segment 2" "segment 3" "segment 4" "segment 5" cols=2 "Start residue" "End residue" +}
-{===>} start_fle_2_1="";
-{===>} end_fle_2_1="";
+{===>} start_fle_2_1="1";
+{===>} end_fle_2_1="11";
 {===>} start_fle_2_2="";
 {===>} end_fle_2_2="";
 {===>} start_fle_2_3="";
@@ -3150,7 +3150,7 @@
 {* Error treshold for restraint violation is defined by error_dih *}
 {+ choice: none all alpha alphabeta +}
 {+ define the error treshold for the restraint violation +}
-{===>} ssdihed=none;
+{===>} ssdihed=alphabeta;
 {===>} error_dih=10;
 
 
@@ -4373,14 +4373,14 @@
 {* Setting for the rigid-body (it0) and semi-flexible refiment (it1) *}
 
 {* number of structures for rigid body docking *}
-{===>} structures_0=1000;
+{===>} structures_0=6000;
        keepstruct_0=&structures_0;
 {* number of structures for refinement *}
-{===>} structures_1=200;
+{===>} structures_1=400;
        keepstruct_1=&structures_1;
        keepstruct_2=&structures_1;
 {* number of structures to be analysed*}
-{===>} anastruc_1=200;
+{===>} anastruc_1=400;
        anastruc_0=&anastruc_1;
        anastruc_2=&anastruc_1;
 
@@ -4465,16 +4465,16 @@
 {===>} emstepstrans=1000;
 
 {* number of MD steps for rigid body high temperature TAD *}
-{===>} initiosteps=500;
+{===>} initiosteps=2000;
 
 {* number of MD steps during first rigid body cooling stage *}
-{===>} cool1_steps=500;
+{===>} cool1_steps=2000;
 
 {* number of MD steps during second cooling stage with flexible side-chains at interface *}
-{===>} cool2_steps=1000;
+{===>} cool2_steps=4000;
 
 {* number of MD steps during third cooling stage with fully flexible interface *}
-{===>} cool3_steps=1000;
+{===>} cool3_steps=4000;
 
 
 {======================= Solvated rigid body docking=======================}
@@ -4551,7 +4551,7 @@
 
 {* number of structures for the explicit solvent refinement *}
 {* the n best structures will be refined                    *}
-{===>} waterrefine=200;
+{===>} waterrefine=400;
        structures_2=&waterrefine;
 
 {* number of steps for heating phase (100, 200, 300K)?      *}
@@ -4652,10 +4652,10 @@
 
 {* Clustering method (RMSD or Fraction of Common Contacts (FCC)) *}
 {+ choice: "RMSD" "FCC" +}
-{===>} clust_meth="FCC";
+{===>} clust_meth="RMSD";
 
 {* RMSD cutoff for clustering? (Recommended values: RMSD 7.5, FCC 0.60) *}
-{===>} clust_cutoff=0.60;
+{===>} clust_cutoff=5.0;
 
 {* Minimum cluster size? *}
 {===>} clust_size=4;
