{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "V28",
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "TPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# HADDOCK3 tutorial on modelling an antibody-antigen complex"
      ],
      "metadata": {
        "id": "uRn70mwIKlvB"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Introduction\n",
        "\n",
        "This tutorial demonstrates the use of the new modular HADDOCK3 version for predicting\n",
        "the structure of an antibody-antigen complex using knowledge of the hypervariable loops\n",
        "on the antibody (i.e., the most basic knowledge) and epitope information identified from NMR experiments for the antigen to guide the docking.\n",
        "# New Section\n",
        "An antibody is a large protein that generally works by attaching itself to an antigen,\n",
        "which is a unique site of the pathogen. The binding harnesses the immune system to directly\n",
        "attack and destroy the pathogen. Antibodies can be highly specific while showing low immunogenicity (i.e. the ability to provoke an immune response),\n",
        "which is achieved by their unique structure. **The fragment crystallizable region (Fc region)**\n",
        "activates the immune response and is species-specific, i.e. the human Fc region should not\n",
        "induce an immune response in humans. **The fragment antigen-binding region (Fab region**)\n",
        "needs to be highly variable to be able to bind to antigens of various nature (high specificity).\n",
        "In this tutorial, we will concentrate on the terminal **variable domain (Fv)** of the Fab region.\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "  <img src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/antibody_described.png\">\n",
        "</figure>\n",
        "\n",
        "The small part of the Fab region that binds the antigen is called **paratope**. The part of the antigen\n",
        "that binds to an antibody is called **epitope**. The paratope consists of six highly flexible loops,\n",
        "known as **complementarity-determining regions (CDRs)** or hypervariable loops whose sequence\n",
        "and conformation are altered to bind to different antigens. CDRs are shown in red in the figure below:\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "  <img src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/CDRs.png\">\n",
        "</figure>\n",
        "\n",
        "In this tutorial we will be working with Interleukin-1β (IL-1β)\n",
        "(PDB code [4I1B](https://www.ebi.ac.uk/pdbe/entry/pdb/4i1b)) as an antigen\n",
        "and its highly specific monoclonal antibody gevokizumab\n",
        "(PDB code [4G6K](https://www.ebi.ac.uk/pdbe/entry/pdb/4g6k))\n",
        "(PDB code of the complex [4G6M](https://www.ebi.ac.uk/pdbe/entry/pdb/4g6m)).\n"
      ],
      "metadata": {
        "id": "Rh-vaM_9woqM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "# A brief introduction to HADDOCK3\n",
        "\n",
        "HADDOCK3 is the next generation integrative modelling software in the\n",
        "long-lasting HADDOCK project. It represents a complete rethinking and rewriting\n",
        "of the HADDOCK2.X series, implementing a new way to interact with HADDOCK and\n",
        "offering new features to users who can now define custom workflows.\n",
        "\n",
        "In the previous HADDOCK2.x versions, users had access to a highly\n",
        "parameterisable yet rigid simulation pipeline composed of three steps:\n",
        "`rigid-body docking (it0)`, `semi-flexible refinement (it1)`, and `final refinement (itw)`.\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "<img width=\"75%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/HADDOCK2-stages.png\">\n",
        "</figure>\n",
        "\n",
        "In HADDOCK3, users have the freedom to configure docking workflows into\n",
        "functional pipelines by combining the different HADDOCK3 modules, thus\n",
        "adapting the workflows to their projects. HADDOCK3 has therefore developed to\n",
        "truthfully work like a puzzle of many pieces (simulation modules) that users can\n",
        "combine freely. To this end, the “old” HADDOCK machinery has been modularized,\n",
        "and several new modules added, including third-party software additions. As a\n",
        "result, the modularization achieved in HADDOCK3 allows users to duplicate steps\n",
        "within one workflow (e.g., to repeat twice the `it1` stage of the HADDOCK2.x\n",
        "rigid workflow).\n",
        "\n",
        "Note that, for simplification purposes, at this time, not all functionalities of\n",
        "HADDOCK2.x have been ported to HADDOCK3, which does not (yet) support NMR RDC,\n",
        "PCS and diffusion anisotropy restraints, cryo-EM restraints and coarse-graining.\n",
        "Any type of information that can be converted into ambiguous interaction\n",
        "restraints can, however, be used in HADDOCK3, which also supports the\n",
        "*ab initio* docking modes of HADDOCK.\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "<img width=\"75%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/HADDOCK3-workflow-scheme.png\">\n",
        "</figure>\n",
        "\n",
        "To keep HADDOCK3 modules organized, we catalogued them into several\n",
        "categories. However, there are no constraints on piping modules of different\n",
        "categories.\n",
        "\n",
        "The main module categories are \"topology\", \"sampling\", \"refinement\",\n",
        "\"scoring\", and \"analysis\". There is no limit to how many modules can belong to a\n",
        "category. Modules are added as developed, and new categories will be created\n",
        "if/when needed. You can access the HADDOCK3 documentation page for the list of\n",
        "all categories and modules. Below is a summary of the available modules:\n",
        "\n",
        "* **Topology modules**\n",
        "    * `topoaa`: *generates the all-atom topologies for the CNS engine.*\n",
        "* **Sampling modules**\n",
        "    * `rigidbody`: *Rigid body energy minimization with CNS (`it0` in haddock2.x).*\n",
        "    * `lightdock`: *Third-party glow-worm swam optimization docking software.*\n",
        "* **Model refinement modules**\n",
        "    * `flexref`: *Semi-flexible refinement using a simulated annealing protocol through molecular dynamics simulations in torsion angle space (`it1` in haddock2.x).*\n",
        "    * `emref`: *Refinement by energy minimisation (`itw` EM only in haddock2.4).*\n",
        "    * `mdref`: *Refinement by a short molecular dynamics simulation in explicit solvent (`itw` in haddock2.X).*\n",
        "* **Scoring modules**\n",
        "    * `emscoring`: *scoring of a complex performing a short EM (builds the topology and all missing atoms).*\n",
        "    * `mdscoring`: *scoring of a complex performing a short MD in explicit solvent + EM (builds the topology and all missing atoms).*\n",
        "* **Analysis modules**\n",
        "    * `alascan`: *Performs a systematic (or user-define) alanine scanning mutagenesis of interface residues.*\n",
        "    * `caprieval`: *Calculates CAPRI metrics (i-RMSD, l-RMSD, Fnat, DockQ) with respect to the top-scoring model or reference structure if provided.*\n",
        "    * `clustfcc`: *Clusters models based on the fraction of common contacts (FCC)*\n",
        "    * `clustrmsd`: *Clusters models based on pairwise RMSD matrix calculated with the `rmsdmatrix` module.*\n",
        "    * `contactmap`: *Generate contact matrices of both intra- and intermolecular contacts and a chordchart of intermolecular contacts.*\n",
        "    * `rmsdmatrix`: *Calculates the pairwise RMSD matrix between all the models generated in the previous step.*\n",
        "    * `ilrmsdmatrix`: *Calculates the pairwise interface-ligand-RMSD (il-RMSD) matrix between all the models generated in the previous step.*\n",
        "    * `seletop`: *Selects the top N models from the previous step.*\n",
        "    * `seletopclusts`: *Selects the top N clusters from the previous step.*\n",
        "\n",
        "The HADDOCK3 workflows are defined in simple configuration text files, similar to the TOML format but with extra features.\n",
        "Contrary to HADDOCK2.X which follows a rigid (yet highly parameterisable)\n",
        "procedure, in HADDOCK3, you can create your own simulation workflows by\n",
        "combining a multitude of independent modules that perform specialized tasks.\n",
        "\n"
      ],
      "metadata": {
        "id": "V5b-aupVx27d"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "# Software and data setup\n",
        "\n",
        "In order to follow this tutorial we will start Jupyter session in Colab or other resources, install haddock3 and download the data for the tutorial.\n",
        "\n",
        "To run this tutorial on Colab we recommand changing the runtime type to v2-8 TPU, which will give you access to up to 96 cores. The haddock3 docking run described in this tutorial should complete in about 15 minutes.\n",
        "\n",
        "For this follow the following steps:"
      ],
      "metadata": {
        "id": "bs6xG1w1DTaW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 1. Install haddock3, BioPython and py3Dmol\n",
        "#@markdown Execute this cell to download the required packages.\n",
        "!pip install haddock3==2025.8.0 --quiet\n",
        "!pip install py3Dmol --quiet\n",
        "!pip install BioPython --quiet"
      ],
      "metadata": {
        "collapsed": true,
        "id": "91XeT81tMPli"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2. Initiate custom Python functions for the 3D visualisation and alignement of the docking models { display-mode: \"form\" }\n",
        "#@markdown Execute this cell to initiate align_full function to enable its use later in the tutorial.\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "from Bio.PDB import PDBParser, PDBIO, Superimposer\n",
        "from Bio.PDB.Structure import Structure\n",
        "from io import StringIO\n",
        "import numpy as np\n",
        "\n",
        "def load_pdb_file(file_path):\n",
        "  if not os.path.exists(file_path):\n",
        "    print(f\"Error: File not found at {file_path}\")\n",
        "    return None\n",
        "\n",
        "  if file_path.endswith('.gz'):\n",
        "    with gzip.open(file_path, 'rt') as f:\n",
        "      return f.read()\n",
        "  else:\n",
        "    with open(file_path, 'r') as f:\n",
        "      return f.read()\n",
        "\n",
        "\n",
        "def pdb_string_to_structure(pdb_string, structure_id):\n",
        "  parser = PDBParser(QUIET=True)\n",
        "  pdb_io = StringIO(pdb_string)\n",
        "  structure = parser.get_structure(structure_id, pdb_io)\n",
        "  return structure\n",
        "\n",
        "\n",
        "def structure_to_pdb_string(structure):\n",
        "  pdb_io = PDBIO()\n",
        "  pdb_io.set_structure(structure)\n",
        "  output = StringIO()\n",
        "  pdb_io.save(output)\n",
        "  return output.getvalue()\n",
        "\n",
        "# Full alignement - user-given chains used\n",
        "def align_full(pdb_path1, pdb_path2, chains=['A', 'B'], width=800, height=600,\n",
        "                       model1_colors={'A': 'red', 'B': 'orange'},\n",
        "                       model2_colors={'A': 'blue', 'B': 'green'},\n",
        "                       atom_types=['CA'], show_labels=False, show_per_chain_rmsd=True):\n",
        "    \"\"\"\n",
        "    Align two protein structures using all specified chains and visualize with py3Dmol.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    pdb_path1 : str\n",
        "        Path to the first (reference) PDB file\n",
        "    pdb_path2 : str\n",
        "        Path to the second PDB file to align to the first\n",
        "    chains : list, default ['A', 'B']\n",
        "        List of chain IDs to include in alignment\n",
        "    width : int, default 800\n",
        "        Viewer width in pixels\n",
        "    height : int, default 600\n",
        "        Viewer height in pixels\n",
        "    model1_colors : dict, default {'A': 'red', 'B': 'orange'}\n",
        "        Colors for chains in model 1\n",
        "    model2_colors : dict, default {'A': 'blue', 'B': 'green'}\n",
        "        Colors for chains in model 2\n",
        "    atom_types : list, default ['CA']\n",
        "        Atom types to use for alignment (['CA'] or ['CA', 'CB', 'N', 'C'])\n",
        "    show_labels : bool, default True\n",
        "        Whether to show descriptive labels\n",
        "    show_per_chain_rmsd : bool, default True\n",
        "        Whether to calculate and display per-chain RMSD values\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    py3Dmol.view object\n",
        "\n",
        "    Example:\n",
        "    --------\n",
        "    align_full_molecule('model1.pdb.gz', 'model2.pdb.gz')\n",
        "    \"\"\"\n",
        "    def get_atoms_from_chains(structure, chain_ids, atom_types):\n",
        "        atoms = []\n",
        "        chain_info = {}\n",
        "\n",
        "        for model in structure:\n",
        "            for chain_id in chain_ids:\n",
        "                if chain_id in model:\n",
        "                    chain = model[chain_id]\n",
        "                    chain_atoms = []\n",
        "                    for residue in chain:\n",
        "                        for atom_type in atom_types:\n",
        "                            if atom_type in residue:\n",
        "                                atoms.append(residue[atom_type])\n",
        "                                chain_atoms.append(residue[atom_type])\n",
        "                    chain_info[chain_id] = len(chain_atoms)\n",
        "\n",
        "        return atoms, chain_info\n",
        "\n",
        "    # Create viewer\n",
        "    view = py3Dmol.view(width=width, height=height)\n",
        "\n",
        "    # Load PDB files\n",
        "    model_1_data = load_pdb_file(pdb_path1)\n",
        "    model_2_data = load_pdb_file(pdb_path2)\n",
        "\n",
        "    if not (model_1_data and model_2_data):\n",
        "        print(\"Failed to load one or both PDB files\")\n",
        "        return view, None, {}\n",
        "\n",
        "    overall_rmsd = None\n",
        "    per_chain_rmsd = {}\n",
        "\n",
        "    try:\n",
        "        # Parse structures\n",
        "        struct1 = pdb_string_to_structure(model_1_data, \"model1\")\n",
        "        struct2 = pdb_string_to_structure(model_2_data, \"model2\")\n",
        "\n",
        "        # Get atoms from all specified chains\n",
        "        atoms_1, chain_info_1 = get_atoms_from_chains(struct1, chains, atom_types)\n",
        "        atoms_2, chain_info_2 = get_atoms_from_chains(struct2, chains, atom_types)\n",
        "\n",
        "        print(f\"Atoms for alignment - Model 1: {chain_info_1}, Total: {len(atoms_1)}\")\n",
        "        print(f\"Atoms for alignment - Model 2: {chain_info_2}, Total: {len(atoms_2)}\")\n",
        "        print(f\"Model 1: \" + \"; \".join([f\"chain {chain} in {color}\" for chain, color in model1_colors.items() if chain in chains]))\n",
        "        print(f\"Model 2: \" + \"; \".join([f\"chain {chain} in {color}\" for chain, color in model2_colors.items() if chain in chains]))\n",
        "\n",
        "        if len(atoms_1) > 0 and len(atoms_2) > 0:\n",
        "            # Align using all atoms from specified chains\n",
        "            min_atoms = min(len(atoms_1), len(atoms_2))\n",
        "            ref_atoms = atoms_1[:min_atoms]\n",
        "            alt_atoms = atoms_2[:min_atoms]\n",
        "\n",
        "            print(f\"Using {min_atoms} atom pairs for alignment\")\n",
        "\n",
        "            # Perform superimposition\n",
        "            sup = Superimposer()\n",
        "            sup.set_atoms(ref_atoms, alt_atoms)\n",
        "            overall_rmsd = sup.rms\n",
        "\n",
        "            print(f\"Whole molecule alignment RMSD: {overall_rmsd:.3f} Å\")\n",
        "\n",
        "            # Apply transformation to all atoms in structure 2\n",
        "            sup.apply(struct2.get_atoms())\n",
        "\n",
        "            # Calculate per-chain RMSD if requested\n",
        "            if show_per_chain_rmsd:\n",
        "                for chain_id in chains:\n",
        "                    try:\n",
        "                        chain_atoms_1, _ = get_atoms_from_chains(struct1, [chain_id], atom_types)\n",
        "                        chain_atoms_2, _ = get_atoms_from_chains(struct2, [chain_id], atom_types)\n",
        "                        if len(chain_atoms_1) > 0 and len(chain_atoms_2) > 0:\n",
        "                            min_chain = min(len(chain_atoms_1), len(chain_atoms_2))\n",
        "                            sup_chain = Superimposer()\n",
        "                            sup_chain.set_atoms(chain_atoms_1[:min_chain], chain_atoms_2[:min_chain])\n",
        "                            per_chain_rmsd[chain_id] = sup_chain.rms\n",
        "                            print(f\"Chain {chain_id} RMSD: {sup_chain.rms:.3f} Å\")\n",
        "                    except Exception as e:\n",
        "                        print(f\"Could not calculate RMSD for chain {chain_id}: {e}\")\n",
        "\n",
        "            # Convert back to PDB strings\n",
        "            aligned_pdb_1 = structure_to_pdb_string(struct1)\n",
        "            aligned_pdb_2 = structure_to_pdb_string(struct2)\n",
        "\n",
        "            # Add models to viewer\n",
        "            view.addModel(aligned_pdb_1, 'pdb')\n",
        "            view.addModel(aligned_pdb_2, 'pdb')\n",
        "\n",
        "        else:\n",
        "            print(\"Could not find sufficient atoms for alignment, adding original models\")\n",
        "            view.addModel(model_1_data, 'pdb')\n",
        "            view.addModel(model_2_data, 'pdb')\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Alignment failed: {e}\")\n",
        "        view.addModel(model_1_data, 'pdb')\n",
        "        view.addModel(model_2_data, 'pdb')\n",
        "\n",
        "    # Apply styling\n",
        "    view.setStyle({'model': 0}, {'cartoon': {}})\n",
        "    view.setStyle({'model': 1}, {'cartoon': {}})\n",
        "\n",
        "    for chain, color in model1_colors.items():\n",
        "        if chain in chains:\n",
        "            view.addStyle({'model': 0, 'chain': chain},\n",
        "                         {'cartoon': {'color': color, 'opacity': 0.9}})\n",
        "\n",
        "    for chain, color in model2_colors.items():\n",
        "        if chain in chains:\n",
        "            view.addStyle({'model': 1, 'chain': chain},\n",
        "                         {'cartoon': {'color': color, 'opacity': 0.6}})\n",
        "\n",
        "    # Add labels if requested\n",
        "    if show_labels:\n",
        "        view.addLabel(\"Model 1 (Reference)\", {'position': {'x': -20, 'y': 20, 'z': 0},\n",
        "                                             'backgroundColor': 'darkred', 'fontColor': 'white'})\n",
        "        view.addLabel(\"Model 2 (Aligned)\", {'position': {'x': 20, 'y': 20, 'z': 0},\n",
        "                                           'backgroundColor': 'darkgreen', 'fontColor': 'white'})\n",
        "        view.addLabel(f\"Full Molecule Alignment\", {'position': {'x': 0, 'y': -20, 'z': 0},\n",
        "                                                  'backgroundColor': 'navy', 'fontColor': 'white'})\n",
        "        if overall_rmsd:\n",
        "            view.addLabel(f\"Overall RMSD: {overall_rmsd:.3f} Å\", {'position': {'x': 0, 'y': -30, 'z': 0},\n",
        "                                                                 'backgroundColor': 'purple', 'fontColor': 'white'})\n",
        "\n",
        "            # Add per-chain RMSD labels\n",
        "            y_offset = -40\n",
        "            for chain_id, rmsd_val in per_chain_rmsd.items():\n",
        "                view.addLabel(f\"Chain {chain_id}: {rmsd_val:.3f} Å\",\n",
        "                             {'position': {'x': 0, 'y': y_offset, 'z': 0},\n",
        "                              'backgroundColor': 'gray', 'fontColor': 'white', 'fontSize': 10})\n",
        "                y_offset -= 8\n",
        "\n",
        "    view.zoomTo()\n",
        "\n",
        "    return view"
      ],
      "metadata": {
        "id": "mltmuSHiMqhi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3. Download the tutorial data\n",
        "#@markdown We are providing pre-processed PDB files for docking and analysis (but the preprocessing of those files will also be explained in this tutorial). The files have been processed to facilitate their use in HADDOCK and to allow comparison with the known reference structure of the complex. Execute this cell to download and decompress the data.\n",
        "!wget https://surfdrive.surf.nl/files/index.php/s/LdmwnxWFkI9l75p/download -O HADDOCK3-antibody-antigen-notebook.zip\n",
        "!unzip -q HADDOCK3-antibody-antigen-notebook.zip\n",
        "import os\n",
        "from pathlib import Path\n",
        "BASE_DIR = Path.cwd()\n",
        "PROJECT_NAME = 'HADDOCK3-antibody-antigen-notebook'\n",
        "PROJECT_DIR = BASE_DIR / PROJECT_NAME\n",
        "os.chdir(PROJECT_DIR)"
      ],
      "metadata": {
        "id": "87eBZVJD_uJE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Unziping the file will create the HADDOCK3-antibody-antigen directory which should contain the following directories and files:\n",
        "\n",
        "* pdbs: a directory containing the pre-processed PDB files\n",
        "* restraints: a directory containing the interface information and the corresponding restraint files for HADDOCK3\n",
        "* runs: a directory containing pre-calculated results\n",
        "* scripts: a directory containing various scripts used in this tutorial\n",
        "* workflows: a directory containing configuration file examples for HADDOCK3\n"
      ],
      "metadata": {
        "id": "YRXMH2r-_5-v"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "---\n",
        "\n",
        "# Preparing PDB files for docking\n",
        "\n",
        "In this section we will prepare the PDB files of the antibody and antigen for docking.\n",
        "Crystal structures of both the antibody and the antigen in their free forms are available from the\n",
        "[PDBe database](https://www.pdbe.org).\n",
        "\n",
        "*__Important:__ For a docking run with HADDOCK, each molecule should consist of a single chain with non-overlapping residue numbering within the same chain.\n",
        "\n",
        "As an antibody consists of two chains (L+H), we will have to prepare it for use in HADDOCK. For this we will be making use of `pdb-tools` from the command line.\n",
        "\n",
        "_**Note**_ that `pdb-tools` is also available as a [web service](https://wenmr.science.uu.nl/pdbtools/).\n"
      ],
      "metadata": {
        "id": "qEcAhjfK1BxK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<br>\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "## Preparing the antibody structure\n",
        "\n",
        "Using PDB-tools we will download the unbound structure of the antibody from the PDB database (the PDB ID is [4G6K](https://www.ebi.ac.uk/pdbe/entry/pdb/4g6k))\n",
        "and then process it to have a unique chain ID (A) and non-overlapping residue numbering by renumbering the merged pdb (starting from 1). For this we will concatenate the following PDB-tools commands:\n",
        "\n",
        "1. fetch the PDB entry from the PDB database (`pdb_fetch`)\n",
        "2. clean the PDB file (`pdb_tidy`)\n",
        "3. select the chain (`pdb_selchain`),\n",
        "4. remove any hetero atoms from the structure (e.g. crystal waters, small molecules from the crystallisation buffer and such) (`pdb_delhetatm`),\n",
        "5. fix residue numbering insertion in the HV loops (often occuring in antibodies - see note below) (`pdb_fixinsert`)\n",
        "6. remove any possible side-chain duplication (can be present in high-resolution crystal structures in case of multiple conformations of some side chains) (`pdb_selaltloc`)\n",
        "7. keep only the coordinates lines (`pdb_keepcoord`),\n",
        "8. select only the variable domain (FV) of the antibody (to reduce computing time) (`pdb_selres`)\n",
        "9. clean the PDB file (`pdb_tidy`)\n",
        "\n",
        "**_Note_** that the `pdb_tidy -strict` commands cleans the PDB file, add TER statements only between different chains).\n",
        "Without the -strict option TER statements would be added between each chain break (e.g. missing residues), which should be avoided.\n",
        "\n",
        "**_Note_**: An important part for antibodies is the `pdb_fixinsert` command that fixes the residue numbering of the HV loops:\n",
        "Antibodies often follow the [Chothia numbering scheme](https://pubmed.ncbi.nlm.nih.gov/9367782/?otool=inluulib)\n",
        "and insertions created by this numbering scheme (e.g. 82A, 82B, 82C) cannot be processed by HADDOCK directly\n",
        "(if not done those residues will not be considered resulting effectively in a break in the loop).\n",
        "As such, renumbering is necessary before starting the docking.\n",
        "\n",
        "\n",
        "This can be done from the command line with:\n",
        "\n",
        "1. Fetch and process chain H\n",
        "```\n",
        "pdb_fetch 4G6K | pdb_tidy -strict | pdb_selchain -H | pdb_delhetatm | pdb_fixinsert | pdb_selaltloc | pdb_keepcoord | pdb_selres -1:120 | pdb_tidy -strict > 4G6K_H.pdb\n",
        "```\n",
        "\n",
        "2. Fetch and process chain L\n",
        "```\n",
        "pdb_fetch 4G6K | pdb_tidy -strict | pdb_selchain -L | pdb_delhetatm | pdb_fixinsert | pdb_selaltloc | pdb_keepcoord | pdb_selres -1:107 | pdb_tidy -strict > 4G6K_L.pdb\n",
        "```\n",
        "\n",
        "3. combine the heavy and light chain into one, renumbering the residues starting at 1 to avoid overlap in residue numbering between the chains and assigning a unique chainID/segID:\n",
        "\n",
        "```\n",
        "pdb_merge 4G6K_H.pdb 4G6K_L.pdb | pdb_reres -1 | pdb_chain -A | pdb_chainxseg | pdb_tidy -strict > 4G6K_clean.pdb\n",
        "```\n",
        "\n",
        "_**Note**_ The ready-to-use file can be found in the `pdbs` directory of the provided tutorial data.\n"
      ],
      "metadata": {
        "id": "DbUC2WBL4ChR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pdb_fetch 4G6K | pdb_tidy -strict | pdb_selchain -H | pdb_delhetatm | pdb_fixinsert | pdb_selaltloc | pdb_keepcoord | pdb_selres -1:120 | pdb_tidy -strict > 4G6K_H.pdb\n",
        "!pdb_fetch 4G6K | pdb_tidy -strict | pdb_selchain -L | pdb_delhetatm | pdb_fixinsert | pdb_selaltloc | pdb_keepcoord | pdb_selres -1:10 | pdb_tidy -strict > 4G6K_L.pdb\n",
        "!pdb_merge 4G6K_H.pdb 4G6K_L.pdb | pdb_reres -1 | pdb_chain -A | pdb_chainxseg | pdb_tidy -strict > 4G6K_clean.pdb"
      ],
      "metadata": {
        "id": "qVAXTKWv4p22"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "---\n",
        "\n",
        "\n",
        "## Preparing the antigen structure\n",
        "\n",
        "Using PDB-tools, we will now download the unbound structure of Interleukin-1β from the PDB database (the PDB ID is [4I1B](https://www.ebi.ac.uk/pdbe/entry/pdb/4i1b)),\n",
        "remove the hetero atoms and then process it to assign it chainID B.\n",
        "\n",
        "*__Important__: Each molecule given to HADDOCK in a docking scenario must have a unique chainID/segID.*\n"
      ],
      "metadata": {
        "id": "Uflar20y4Nhy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pdb_fetch 4I1B | pdb_tidy -strict | pdb_delhetatm  | pdb_selaltloc | pdb_keepcoord | pdb_chain -B | pdb_chainxseg | pdb_tidy -strict > 4I1B_clean.pdb"
      ],
      "metadata": {
        "id": "sPo5QBve5HCc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "pV3xfNxa4oUH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "# Defining restraints for docking\n",
        "\n",
        "Before setting up the docking, we first need to generate distance restraint files in a format suitable for HADDOCK.\n",
        "HADDOCK uses [CNS][link-cns]{:target=\"_blank\"} as computational engine.\n",
        "A description of the format for the various restraint types supported by HADDOCK can be found in our [Nature Protocol 2024][nat-pro]{:target=\"_blank\"} paper, Box 1.\n",
        "\n",
        "Distance restraints are defined as follows:\n",
        "\n",
        "<pre>\n",
        "assign (selection1) (selection2) distance, lower-bound correction, upper-bound correction\n",
        "</pre>\n",
        "\n",
        "The lower limit for the distance is calculated as: distance minus lower-bound correction\n",
        "and the upper limit as: distance plus upper-bound correction.\n",
        "\n",
        "The syntax for the selections can combine information about:\n",
        "\n",
        "* chainID - `segid` keyword\n",
        "* residue number - `resid` keyword\n",
        "* atom name - `name` keyword.\n",
        "\n",
        "Other keywords can be used in various combinations of OR and AND statements. Please refer for that to the [online CNS manual][link-cns]{:target=\"_blank\"}.\n",
        "\n",
        "E.g.: a distance restraint between the CB carbons of residues 10 and 200 in chains A and B with an\n",
        "allowed distance range between 10Å and 20Å would be defined as follows:\n",
        "\n",
        "<pre>\n",
        "assign (segid A and resid 10 and name CB) (segid B and resid 200 and name CB) 20.0 10.0 0.0\n",
        "</pre>\n",
        "\n",
        "__*Can you think of a different way of defining the distance and lower and upper corrections while maintaining the same allowed range?*__\n",
        "\n"
      ],
      "metadata": {
        "id": "xOYKbdyH4m6n"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "### Identifying the paratope of the antibody\n",
        "\n",
        "Nowadays several computational tools can identify the paratope (the residues of the hypervariable loops involved in binding) from the provided antibody sequence.\n",
        "In this tutorial, we are providing you with the corresponding list of residue obtained using [ProABC-2](https://github.com/haddocking/proabc-2).\n",
        "ProABC-2 uses a convolutional neural network to identify not only residues which are located in the paratope region but also the nature of interactions they are most likely involved in (hydrophobic or hydrophilic).\n",
        "The work is described in [Ambrosetti, *et al* Bioinformatics, 2020](https://academic.oup.com/bioinformatics/article/36/20/5107/5873593).\n",
        "\n",
        "The corresponding paratope residues (those with either an overall probability >= 0.4 or a probability for hydrophobic or hydrophilic > 0.3) are:\n",
        "\n",
        "<pre>\n",
        "31,32,33,34,35,52,54,55,56,100,101,102,103,104,105,106,151,152,169,170,173,211,212,213,214,216\n",
        "</pre>\n",
        "\n",
        "The numbering corresponds to the numbering of the `4G6K_clean.pdb` PDB file.\n",
        "\n",
        "Let us visualize those onto the 3D structure. For this execute the next cell.\n"
      ],
      "metadata": {
        "id": "PlNFKNg56PDW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualize the antibody paratope\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"pdbs\" / \"4G6K_clean.pdb\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {'color': 'white'}})\n",
        "    view.addStyle({'resi': [31,32,33,34,35,52,54,55,56,100,101,102,103,104,105,106,151,152,169,170,173,211,212,213,214,216]},{'cartoon':{'color':'red'}})\n",
        "    view.addStyle({'resi': [31,32,33,34,35,52,54,55,56,100,101,102,103,104,105,106,151,152,169,170,173,211,212,213,214,216]},{'stick':{'color':'red'}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "A_FbJyrn7MsW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "__*Do the identified paratope residues form a well-defined patch on the surface?*__\n",
        "\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>See surface view of the paratope</i></b>\n",
        "  </summary>\n",
        "  <figure style=\"text-align: center;\">\n",
        "    <img width=\"60%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/antibody-paratope.png\">\n",
        "  </figure>\n",
        "  <br>\n",
        "</details>\n",
        "\n"
      ],
      "metadata": {
        "id": "Ky_OhAu8EjFI"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "## Identifying the epitope of the antigen\n",
        "\n",
        "The article describing the crystal structure of the antibody-antigen complex we are modeling also reports experimental NMR chemical shift titration experiments\n",
        "to map the binding site of the antibody (gevokizumab) on Interleukin-1β.\n",
        "The residues affected by binding are listed in Table 5 of [Blech et al. JMB 2013](https://dx.doi.org/10.1016/j.jmb.2012.09.021):\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "  <img width=\"40%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/Table5-Blech.png\">\n",
        "</figure>\n",
        "\n",
        "The list of binding site (epitope) residues identified by NMR is:\n",
        "\n",
        "<pre>\n",
        "72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117\n",
        "</pre>\n",
        "\n",
        "We will now visualize the epitope on Interleukin-1β.\n"
      ],
      "metadata": {
        "id": "ZW8e_2qK7KzO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualize the NMR-defined epitope\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = './pdbs/4I1B_clean.pdb'\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if pdb_file_path.endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {'color': 'white'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117]},{'cartoon':{'color':'red'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117]},{'stick':{'color':'red'}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "X-ISDzkrFK9-",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Inspect the surface.\n",
        "\n",
        "The answer to that question should be yes, but we can see some residues not colored that might also be involved in the binding - there are some white spots around/in the red surface.\n",
        "\n",
        "\n",
        "__*Do the identified residues form a well-defined patch on the surface?*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>See surface view of the epitope identified by NMR</i></b>\n",
        "  </summary>\n",
        "  <figure style=\"text-align: center;\">\n",
        "    <img width=\"60%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/antigen-epitope.png\">\n",
        "  </figure>\n",
        "  <br>\n",
        "</details>\n"
      ],
      "metadata": {
        "id": "mEJWwrhSFLge"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "## Defining surface neighbours as passive residues\n",
        "\n",
        "In HADDOCK, we are dealing with potentially incomplete binding sites by defining surface neighbors as `passive` residues.\n",
        "These passive residues are added in the definition of the interface but do not incur any energetic penalty if they are not part of the binding site in the final models.\n",
        "In contrast, residues defined as active (typically the identified or predicted binding site residues) will incur an energetic penalty.\n",
        "When using the HADDOCK2.x webserver, `passive` residues will be automatically defined.\n",
        "Here, since we are using a local version, we need to define those manually.\n",
        "\n",
        "This can easily be done using a haddock3 command line tool in the following way:\n",
        "\n",
        "```\n",
        "haddock3-restraints passive_from_active 4I1B_clean.pdb 72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117\n",
        "```\n",
        "\n",
        "The command prints a list of solvent accessible passive residues, which you should save to a file for further use.\n"
      ],
      "metadata": {
        "id": "RAM58g7kHXTS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!haddock3-restraints passive_from_active 4I1B_clean.pdb 72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117"
      ],
      "metadata": {
        "id": "i4f1--Xx5wHJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualise the NMR-defined epitope with surface neighbours\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"pdbs\" / \"4I1B_clean.pdb\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {'color': 'white'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117]},{'cartoon':{'color':'red'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117]},{'stick':{'color':'red'}})\n",
        "    view.addStyle({'resi': [3,24,46,47,48,50,66,76,77,79,80,82,86,87,88,91,93,95,118,119,120]},{'cartoon':{'color':'green'}})\n",
        "    view.addStyle({'resi': [3,24,46,47,48,50,66,76,77,79,80,82,86,87,88,91,93,95,118,119,120]},{'stick':{'color':'green'}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "JEXiQha7Hu22"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>See the epitope and passive residues</i></b>\n",
        "  </summary>\n",
        "  <figure style=\"text-align: center;\">\n",
        "    <img width=\"60%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/antigen-active-passive.png\">\n",
        "  </figure>\n",
        "  <br>\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "The NMR-identified residues and their surface neighbors generated with the above command can be used to define ambiguous interactions restraints,\n",
        "either using the NMR identified residues as active in HADDOCK, or combining those with the surface neighbors.\n",
        "\n",
        "The difference between `active` and `passive` residues in HADDOCK is as follows:\n",
        "\n",
        "*__Active residues__*: These residues are \"forced\" to be at the interface. If they are not part of the interface in the final models, an energetic penalty will be applied. The interface in this context is defined by the union of active and passive residues on the partner molecules.\n",
        "\n",
        "*__Passive residues__*: These residues are expected to be at the interface. However, if they are not, no energetic penalty is applied.\n",
        "\n",
        "\n",
        "In general, it is better to be too generous rather than too strict in the definition of passive residues.\n",
        "An important aspect is to filter both the active (the residues identified from your mapping experiment) and passive residues by their solvent accessibility.\n",
        "This is done automatically when using the `haddock3-restraints passive_from_active` command: residues with less that 15% relative solvent accessibility (same cutoff as the default in the HADDOCK server) are discared.\n",
        "This is, however, not a hard limit, and you might consider including even more buried residues if some important chemical group seems solvent accessible from a visual inspection."
      ],
      "metadata": {
        "id": "ZlHYawnkHvZZ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "---\n",
        "\n",
        "## Defining ambiguous restraints\n",
        "\n",
        "Once you have identified your active and passive residues for both molecules, you can proceed with the generation of the ambiguous interaction restraints (AIR) file for HADDOCK.\n",
        "For this you can either make use of our online [Generate Restraints](https://rascar.science.uu.nl/haddock-restraints) web service, entering the list of active and passive residues for each molecule, the chainIDs of each molecule and saving the resulting restraint list to a text file, or use another `haddock3-restraints` sub-command.\n",
        "\n",
        "To use our `haddock3-restraints active_passive_to_ambig` script, you need to\n",
        "create for each molecule a file containing two lines:\n",
        "\n",
        "* The first line corresponds to the list of active residues (numbers separated by spaces)\n",
        "* The second line corresponds to the list of passive residues (numbers separated by spaces).\n",
        "\n",
        "*__Important__*: The file must consist of two lines, but a line can be empty (e.g., if you do not want to define active residues for one molecule).\n",
        "However, there must be at least one set of active residue defined for one of the molecules.\n",
        "\n",
        "\n",
        "* For the antibody we will use the predicted paratope as active and no passive residues defined.\n",
        "The corresponding file can be found in the `restraints` directory as `antibody-paratope.act-pass`:\n",
        "\n",
        "```\n",
        "1 32 33 34 35 52 54 55 56 100 101 102 103 104 105 106 151 152 169 170 173 211 212 213 214 216\n",
        "\n",
        "```\n",
        "\n",
        "* For the antigen we will use the NMR-identified epitope as active and the surface neighbors as passive.\n",
        "The corresponding file can be found in the `restraints` directory as `antigen-NMR-epitope.act-pass`:\n",
        "\n",
        "```\n",
        "72 73 74 75 81 83 84 89 90 92 94 96 97 98 115 116 117\n",
        "3 24 46 47 48 50 66 76 77 79 80 82 86 87 88 91 93 95 118 119 120\n",
        "```\n",
        "\n",
        "Using those two files, we can generate the CNS-formatted Ambiguous Interaction Restraints (AIRs) file with the following command:\n",
        "\n",
        "```\n",
        "haddock3-restraints active_passive_to_ambig restraints/antibody-paratope.act-pass restraints/antigen-NMR-epitope.act-pass --segid-one A --segid-two B > ambig-paratope-NMR-epitope.tbl\n",
        "```\n",
        "\n",
        "This generates a file called `ambig-paratope-NMR-epitope.tbl` that contains the AIRs.\n"
      ],
      "metadata": {
        "id": "QITYB4uUauPD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!haddock3-restraints active_passive_to_ambig restraints/antibody-paratope.act-pass restraints/antigen-NMR-epitope.act-pass --segid-one A --segid-two B > ambig-paratope-NMR-epitope.tbl"
      ],
      "metadata": {
        "id": "jKnFsF5X6NBL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "__*Inspect the generated file and note how the ambiguous distances are defined.*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>View an extract of the AIR file</i></b>\n",
        "  </summary>\n",
        "<pre>\n",
        "assign (resi 31 and segid A)\n",
        "(\n",
        "       (resi 72 and segid B)\n",
        "        or\n",
        "       (resi 73 and segid B)\n",
        "        or\n",
        "       (resi 74 and segid B)\n",
        "        or\n",
        "       (resi 75 and segid B)\n",
        "        or\n",
        "       (resi 81 and segid B)\n",
        "        or\n",
        "       (resi 83 and segid B)\n",
        "        or\n",
        "       (resi 84 and segid B)\n",
        "        or\n",
        "       (resi 89 and segid B)\n",
        "        or\n",
        "       (resi 90 and segid B)\n",
        "        or\n",
        "       (resi 92 and segid B)\n",
        "        or\n",
        "       (resi 94 and segid B)\n",
        "        or\n",
        "       (resi 96 and segid B)\n",
        "        or\n",
        "       (resi 97 and segid B)\n",
        "        or\n",
        "       (resi 98 and segid B)\n",
        "        or\n",
        "       (resi 115 and segid B)\n",
        "        or\n",
        "       (resi 116 and segid B)\n",
        "        or\n",
        "       (resi 117 and segid B)\n",
        "        or\n",
        "       (resi 3 and segid B)\n",
        "        or\n",
        "       (resi 24 and segid B)\n",
        "        or\n",
        "       (resi 46 and segid B)\n",
        "        or\n",
        "       (resi 47 and segid B)\n",
        "        or\n",
        "       (resi 48 and segid B)\n",
        "        or\n",
        "       (resi 50 and segid B)\n",
        "        or\n",
        "       (resi 66 and segid B)\n",
        "        or\n",
        "       (resi 76 and segid B)\n",
        "        or\n",
        "       (resi 77 and segid B)\n",
        "        or\n",
        "       (resi 79 and segid B)\n",
        "        or\n",
        "       (resi 80 and segid B)\n",
        "        or\n",
        "       (resi 82 and segid B)\n",
        "        or\n",
        "       (resi 86 and segid B)\n",
        "        or\n",
        "       (resi 87 and segid B)\n",
        "        or\n",
        "       (resi 88 and segid B)\n",
        "        or\n",
        "       (resi 91 and segid B)\n",
        "        or\n",
        "       (resi 93 and segid B)\n",
        "        or\n",
        "       (resi 95 and segid B)\n",
        "        or\n",
        "       (resi 118 and segid B)\n",
        "        or\n",
        "       (resi 119 and segid B)\n",
        "        or\n",
        "       (resi 120 and segid B)\n",
        ") 2.0 2.0 0.0\n",
        "...\n",
        "</pre>\n",
        "  <br>\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "__*Refering to the way the distance restraints are defined (see above), what is the distance range for the ambiguous distance restraints?*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>See answer</i></b>\n",
        "  </summary>\n",
        "The default distance range for those is between 0 and 2Å, which\n",
        "might seem short but makes senses because of the 1/r^6 summation in the AIR\n",
        "energy function that makes the effective distance to be significantly shorter than the shortest distance entering the sum.\n",
        "<br>\n",
        "<br>\n",
        "The effective distance is calculated as the SUM over all pairwise atom-atom\n",
        "distance combinations between an active residue and all the active+passive on\n",
        "the other molecule: SUM[1/r^6]^(-1/6).\n",
        "</details>\n",
        "<br>\n"
      ],
      "metadata": {
        "id": "NmoVBrWS6U9L"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "## Restraints validation\n",
        "\n",
        "If you modify manually this generated restraint files or create your own, it is possible to quickly check if the format is valid using the following `haddock3-restraints` sub-command:\n",
        "\n",
        "```\n",
        "haddock3-restraints validate_tbl ambig-paratope-NMR-epitope.tbl --silent\n",
        "```\n",
        "\n",
        "No output means that your TBL file is valid.\n",
        "\n",
        "*__Note__* that this only validates the syntax of the restraint file, but does not check if the selections defined in the restraints are actually existing in your input PDB files.\n",
        "\n"
      ],
      "metadata": {
        "id": "QaC1vPD0cEZs"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!haddock3-restraints validate_tbl ambig-paratope-NMR-epitope.tbl --silent"
      ],
      "metadata": {
        "id": "sK6rHGWFDllv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "## Additional restraints for multi-chain proteins\n",
        "\n",
        "As an antibody consists of two separate chains, it is important to define a few distance restraints\n",
        "to keep them together during the high temperature flexible refinement stage of HADDOCK otherwise they might slightly drift appart.\n",
        "This can easily be done using the `haddock3-restraints restrain_bodies` sub-command.\n"
      ],
      "metadata": {
        "id": "6Gk-BhfYcNQH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!haddock3-restraints restrain_bodies pdbs/4G6K_clean.pdb > antibody-unambig.tbl"
      ],
      "metadata": {
        "id": "ciFeWsstDqIc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The result file contains two CA-CA distance restraints with the exact distance measured between two randomly picked CA atoms pairs:\n",
        "\n",
        "```  \n",
        "assign (segid A and resi 110 and name CA) (segid A and resi 132 and name CA) 26.326 0.0 0.0\n",
        "assign (segid A and resi 97 and name CA) (segid A and resi 204 and name CA) 19.352 0.0 0.0\n",
        "```\n",
        "\n",
        "This file is also provided in the `restraints` directory.\n",
        "\n"
      ],
      "metadata": {
        "id": "9xLBFZLdDqrq"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "# Setting up and running the docking with HADDOCK3\n",
        "\n",
        "Now that we have all required files at hand (PDB and restraints files), it is time to setup our docking protocol.\n",
        "In this tutorial, considering we have rather good information about the paratope and epitope, we will execute a fast HADDOCK3 docking workflow,\n",
        "reducing the non-negligible computational cost of HADDOCK by decreasing the sampling, without impacting too much the accuracy of the resulting models.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "p5uTYVPKxMRa"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "## HADDOCK3 workflow definition\n",
        "\n",
        "The first step is to create a HADDOCK3 configuration file that will define the docking workflow.\n",
        "We will follow a classic HADDOCK workflow consisting of rigid body docking, semi-flexible refinement and final energy minimisation followed by clustering.\n",
        "\n",
        "We will also integrate two analysis modules in our workflow:\n",
        "\n",
        "- `caprieval` will be used at various stages to compare models to either the best scoring model (if no reference is given) or a reference structure, which in our case we have at hand (`pdbs/4G6M_matched.pdb`). This will directly allow us to assess the performance of the protocol. In the absence of a reference, `caprieval` is still usefull to assess the convergence of a run and analyse the results.\n",
        "- `contactmap` added as last module will generate contact matrices of both intra- and intermolecular contacts and a chordchart of intermolecular contacts for each cluster.\n",
        "\n",
        "\n",
        "Our workflow consists of the following modules:\n",
        "\n",
        "1. **`topoaa`**: *Generates the topologies for the CNS engine and builds missing atoms*\n",
        "2. **`rigidbody`**: *Performs rigid body energy minimisation (`it0` in haddock2.x)*\n",
        "3. **`caprieval`**: *Calculates CAPRI metrics (i-RMSD, l-RMSD, Fnat, DockQ) with respect to the top scoring model or reference structure if provided*\n",
        "4. **`seletop`** : *Selects the top X models from the previous module*\n",
        "5. **`flexref`**: *Preforms semi-flexible refinement of the interface (`it1` in haddock2.4)*\n",
        "6. **`caprieval`**\n",
        "7. **`emref`**: *Final refinement by energy minimisation (`itw` EM only in haddock2.4)*\n",
        "8. **`caprieval`**\n",
        "9. **`clustfcc`**: *Clustering of models based on the fraction of common contacts (FCC)*\n",
        "10. **`seletopclusts`**: *Selects the top models of all clusters*\n",
        "11. **`caprieval`**\n",
        "12. **`contactmap`**: *Contacts matrix and a chordchart of intermolecular contacts*\n",
        "\n",
        "\n",
        "The corresponding toml configuration file (provided in `workflows/docking-antibody-antigen.cfg`) looks like:\n",
        "\n",
        "```\n",
        "# ====================================================================\n",
        "# Antibody-antigen docking example with restraints from the antibody\n",
        "# paratope to the NMR-identified epitope on the antigen\n",
        "# ====================================================================\n",
        "\n",
        "# Directory in which the scoring will be done\n",
        "run_dir = \"run1\"\n",
        "\n",
        "# Compute mode\n",
        "mode = \"local\"\n",
        "ncores = 50\n",
        "\n",
        "# Self contained rundir (to avoid problems with long filename paths)\n",
        "self_contained = true\n",
        "\n",
        "# Post-processing to generate statistics and plots\n",
        "postprocess = true\n",
        "clean = true\n",
        "\n",
        "molecules =  [\n",
        "    \"pdbs/4G6K_clean.pdb\",\n",
        "    \"pdbs/4I1B_clean.pdb\"\n",
        "    ]\n",
        "\n",
        "# ====================================================================\n",
        "# Parameters for each stage are defined below, prefer full paths\n",
        "# ====================================================================\n",
        "[topoaa]\n",
        "\n",
        "[rigidbody]\n",
        "# CDR to NMR epitope ambig restraints\n",
        "ambig_fname = \"restraints/ambig-paratope-NMR-epitope.tbl\"\n",
        "# Restraints to keep the antibody chains together\n",
        "unambig_fname = \"restraints/antibody-unambig.tbl\"\n",
        "# Reduced sampling (100 instead of the default of 1000)\n",
        "sampling = 100\n",
        "\n",
        "[caprieval]\n",
        "reference_fname = \"pdbs/4G6M_matched.pdb\"\n",
        "\n",
        "[seletop]\n",
        "# Selection of the top 20 best scoring complexes (instead of the default of 200 for speed purposes for the tutorial)\n",
        "select = 20\n",
        "\n",
        "[flexref]\n",
        "tolerance = 5\n",
        "# CDR to NMR epitope ambig restraints\n",
        "ambig_fname = \"restraints/ambig-paratope-NMR-epitope.tbl\"\n",
        "# Restraints to keep the antibody chains together\n",
        "unambig_fname = \"restraints/antibody-unambig.tbl\"\n",
        "\n",
        "[caprieval]\n",
        "reference_fname = \"pdbs/4G6M_matched.pdb\"\n",
        "\n",
        "[emref]\n",
        "tolerance = 5\n",
        "# CDR to NMR epitope ambig restraints\n",
        "ambig_fname = \"restraints/ambig-paratope-NMR-epitope.tbl\"\n",
        "# Restraints to keep the antibody chains together\n",
        "unambig_fname = \"restraints/antibody-unambig.tbl\"\n",
        "\n",
        "[caprieval]\n",
        "reference_fname = \"pdbs/4G6M_matched.pdb\"\n",
        "\n",
        "[clustfcc]\n",
        "plot_matrix = true\n",
        "\n",
        "[seletopclusts]\n",
        "# Selection of the top 4 best scoring complexes from each cluster\n",
        "top_models = 4\n",
        "\n",
        "[caprieval]\n",
        "reference_fname = \"pdbs/4G6M_matched.pdb\"\n",
        "\n",
        "[contactmap]\n",
        "\n",
        "# ====================================================================\n",
        "\n",
        "```\n",
        "\n",
        "\n",
        "In this case, since we have information for both interfaces we use a low-sampling configuration file, which takes only a small amount of computational resources to run.\n",
        "The initial `sampling` parameter at the rigid-body energy minimization (`rigidbody`) module is set to 100 models, of which only best the 40 are passed to the flexible refinement (`flexref`) module with the `seletop` module.\n",
        "The subsequence flexible refinement (`flexref` module) and energy minimisation (*emref*) modules will use all models passed by the *seletop* module.\n",
        "FCC clustering (`clustfcc`) is then applied to group together models sharing a consistent fraction of the interface contacts.\n",
        "The top 4 models of each cluster are saved to disk (`seletopclusts`).\n",
        "\n",
        "Multiple `caprieval` modules are executed at different stages of the workflow to check how the quality (and rankings) of the models change throughout the protocol.\n",
        "In this case we are providing the known crystal structure of the complex as reference.\n",
        "\n",
        "\n",
        "**_Note_**: For making best use of the available CPU resources it is recommended to adapt the sampling parameter to be a multiple of the number of available cores when running in local mode. For this reason, for the ASEAN HPC school the sampling is set to be a multiple of 48.\n",
        "\n",
        "**_Note_**: In case no reference is available (the usual scenario), the best ranked model is used as reference for each stage.\n",
        "Including `caprieval` at the various stages even when no reference is provided is useful to get the rankings and scores and visualise the results (see Analysis section below).\n",
        "\n",
        "**_Note_**: The default sampling would be 1000 models for `rigidbody` of which 200 are passed to the flexible refinement in `seletop`.\n",
        "As an indication of the computational requirements, the default sampling worflow for this tutorial completes in about 37min using 12 cores on a MaxOSX M2 processor.\n",
        "In comparison, the reduced sampling run (100/40) takes about 7-8min.\n",
        "\n",
        "\n",
        "\n",
        "**_Note_**: To get a list of all possible parameters that can be defined in a specific module (and their default values) you can use the following command:\n",
        "\n",
        "```\n",
        "haddock3-cfg -m <module-name>\n",
        "```\n",
        "Add the `-d` option to get a more detailed description of parameters and use the `-h` option to see a list of arguments and options.\n",
        "\n",
        "Alternatively, you can consult the [developer's guide](https://www.bonvinlab.org/haddock3/), where each parameter of each module is listed along with their default values, short and long descriptions, etc. Navigate to the [Modules](https://www.bonvinlab.org/haddock3/modules/index.html#) and select a module which parameters you want to display.\n",
        "\n",
        "__*In the above workflow we see in three modules a *tolerance* parameter defined. Using the `haddock3-cfg` command try to figure out what this parameter does.*__\n",
        "\n",
        "\n",
        "\n",
        "**_Note_** that, in contrast to HADDOCK2.X, we have much more flexibility in defining our workflow.\n",
        "As an example, we could use this flexibility by introducing a clustering step after the initial rigid-body docking stage, selecting a given number of models per cluster and refining all of those.\n",
        "For an example of this strategy see the section about ensemble docking.\n",
        "\n"
      ],
      "metadata": {
        "id": "8cDpSDysxTJk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "---\n",
        "\n",
        "## Running HADDOCK3\n",
        "\n",
        "To execute the above workflow you have two choices:\n",
        "\n",
        "1. __Open a terminal session__ and simply type (check if the location is correct):\n",
        "\n",
        "```\n",
        "cd HADDOCK3-antibody-antigen-notebook\n",
        "haddock3 workflows/docking-antibody-antigen.cfg >&run.out &\n",
        "```\n",
        "\n",
        "This will run haddock3 in background. In the mean time we can contine with the analysis section using the precalculated results.\n",
        "\n",
        "You can check the status of the run by looking at the `run.out` file.\n",
        "\n",
        "2. __Or execute the next cell__, and wait for its completion to proceed with the analysis (time for a cup of coffee :-) as the run should take about 15 minutes to complete)"
      ],
      "metadata": {
        "id": "grF10NMwdwz1"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!haddock3 workflows/docking-antibody-antigen.cfg"
      ],
      "metadata": {
        "collapsed": true,
        "id": "_ALobZkaA-Xx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "In case the execution was interupted you can restart haddock3 from the last completed stage. Check the content of the run1 directory and find out which stage was the last running and not completed. Then restart haddock3 with as addional option: ``--restart 4`` (for example to restart at stage 4). This can be done from the terminal within the `HADDOCK3-antibody-antigen-notebook` directory.\n",
        "\n",
        "```\n",
        "haddock3 workflows/docking-antibody-antigen.cfg --restart 4\n",
        "```"
      ],
      "metadata": {
        "id": "OiRPjHOkjlge"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "# Analysis of the docking results\n",
        "\n",
        "In case something went wrong with the docking (or simply if you do not want to wait for the results) you can find the following precalculated runs in the `runs` directory:\n",
        "- `run1`: docking run created using the unbound antibody.\n",
        "- `run-energetics`: alanine scanning run (see BONUS1 section)\n",
        "- `run-scoring`: combined scoring of HADDOCK, AF2 and AF3 models  (see BONUS2 section).\n",
        "\n",
        "\n",
        "Once your run has completed - inspect the content of the resulting directory.\n",
        "You will find the various steps (modules) of the defined workflow numbered sequentially starting at 0, e.g.:\n",
        "\n",
        "```\n",
        "> ls run1/\n",
        "     00_topoaa/\n",
        "     01_rigidbody/\n",
        "     02_caprieval/\n",
        "     03_seletop/\n",
        "     04_flexref/\n",
        "     05_caprieval/\n",
        "     06_emref/\n",
        "     07_caprieval/\n",
        "     08_clustfcc/\n",
        "     09_seletopclusts/\n",
        "     10_caprieval/\n",
        "     11_contactmap/\n",
        "     analysis/\n",
        "     data/\n",
        "     log\n",
        "     toppar/\n",
        "     traceback/\n",
        "```\n",
        "\n",
        "In addition, there is a log file (text file) and four additional directories:\n",
        "\n",
        "- the `analysis` directory contains various plots to visualize the results for each caprieval step and a general report (`report.html`) that provides all statistics with various plots. You can open this file in your preferred web browser\n",
        "- the `data` directory contains the input data (PDB and restraint files) for the various modules, as well as an input workflow  (in `configurations` directory)\n",
        "- the `toppar` directory contains the force field topology and parameter files (only present when running in self-contained mode)\n",
        "- the `traceback` directory contains `traceback.tsv`, which links all models to see which model originates from which throughout all steps of the workflow.\n",
        "\n",
        "You can find information about the duration of the run at the bottom of the log file.\n",
        "\n",
        "Each sampling/refinement/selection module will contain PDB files.\n",
        "For example, the `09_seletopclusts` directory contains the selected models from each cluster. The clusters in that directory are numbered based\n",
        "on their rank, i.e. `cluster_1` refers to the top-ranked cluster. Information about the origin of these files can be found in that directory in the `seletopclusts.txt` file.\n",
        "\n",
        "The simplest way to extract ranking information and the corresponding HADDOCK scores is to look at the `XX_caprieval` directories (which is why it is a good idea to have it as the final module, and possibly as intermediate steps). This directory will always contain a `capri_ss.tsv` single model statistics file, which contains the model names, rankings and statistics (score, iRMSD, Fnat, lRMSD, ilRMSD and dockq score). E.g. for `10_caprieval`:\n",
        "\n",
        "```\n",
        "model   md5     caprieval_rank  score   irmsd   fnat    lrmsd   ilrmsd  dockq   rmsd    cluster_id      cluster_ranking model-cluster_ranking   air     angles  bonds   bsa     cdih    coup    dani    desolv  dihe    elec    improper        rdcs    rg      sym     total   vdw     vean    xpcs\n",
        "../../09_seletopclusts/cluster_1_model_1.pdb    -       1       -129.035        1.351   0.862   3.731   2.772   0.751   1.557   2       1       1       111.427 278.615 52.004  1799.440        0.000   0.000   0.000   0.725   2042.510        -437.132        72.945  0.000   0.000   0.000   -379.180     -53.476 0.000   0.000\n",
        "../../09_seletopclusts/cluster_1_model_2.pdb    -       2       -114.462        1.502   0.810   3.387   3.100   0.724   1.498   2       1       2       144.192 281.411 52.711  1880.050        0.000   0.000   0.000   3.131   2041.260        -458.989        75.587  0.000   0.000   0.000   -355.011     -40.214 0.000   0.000\n",
        "../../09_seletopclusts/cluster_1_model_3.pdb    -       3       -114.217        1.296   0.776   4.068   2.974   0.721   1.350   2       1       3       132.033 283.299 51.917  1872.550        0.000   0.000   0.000   8.364   2050.640        -510.529        73.423  0.000   0.000   0.000   -412.174     -33.678 0.000   0.000\n",
        "...\n",
        "```\n",
        "\n",
        "If clustering was performed prior to calling the `caprieval` module, the `capri_ss.tsv` file will also contain information about to which cluster the model belongs to and its ranking within the cluster.\n",
        "\n",
        "The relevant statistics are:\n",
        "\n",
        "* **score**: *the HADDOCK score (arbitrary units)*\n",
        "* **irmsd**: *the interface RMSD, calculated over the interfaces the molecules*\n",
        "* **fnat**: *the fraction of native contacts*\n",
        "* **lrmsd**: *the ligand RMSD, calculated on the ligand after fitting on the receptor (1st component)*\n",
        "* **ilrmsd**: *the interface-ligand RMSD, calculated over the interface of the ligand after fitting on the interface of the receptor (more relevant for small ligands for example)*\n",
        "* **dockq**: *the DockQ score, which is a combination of irmsd, lrmsd and fnat and provides a continuous scale between 1 (exactly equal to reference) and 0*\n",
        "\n",
        "Various other terms are also reported including:\n",
        "\n",
        "* **bsa**: *the buried surface area (in squared angstroms)*\n",
        "* **elec**: *the intermolecular electrostatic energy*\n",
        "* **vdw**: *the intermolecular van der Waals energy*\n",
        "* **desolv**: *the desolvation energy*\n",
        "\n",
        "\n",
        "The iRMSD, lRMSD and Fnat metrics are the ones used in the blind protein-protein prediction experiment [CAPRI](https://www.ebi.ac.uk/msd-srv/capri/) (Critical PRediction of Interactions).\n",
        "\n",
        "In CAPRI the quality of a model is defined as (for protein-protein complexes):\n",
        "\n",
        "* **acceptable model**: i-RMSD < 4Å or l-RMSD < 10Å and Fnat > 0.1 (0.23 < DOCKQ < 0.49)\n",
        "* **medium quality model**: i-RMSD < 2Å or l-RMSD < 5Å and Fnat > 0.3 (0.49 < DOCKQ < 0.8)\n",
        "* **high quality model**: i-RMSD < 1Å or l-RMSD < 1Å and Fnat > 0.5 (DOCKQ > 0.8)\n",
        "\n",
        "__*Based on these CAPRI criteria, what is the quality of the best model listed above (_cluster_1_model_1.pdb_)?*__\n",
        "\n",
        "\n",
        "In case where the `caprieval` module is called after a clustering step, an additional `capri_clt.tsv` file will be present in the directory.\n",
        "This file contains the cluster ranking and score statistics, averaged over the minimum number of models defined for clustering\n",
        "(4 by default), with their corresponding standard deviations. E.g.:\n",
        "\n",
        "```\n",
        "cluster_rank    cluster_id      n       under_eval      score   score_std       irmsd   irmsd_std       fnat    fnat_std        lrmsd   lrmsd_std       dockq   dockq_std       ilrmsd  ilrmsd_std      rmsd    rmsd_std        air     air_std bsa     bsa_std desolv  desolv_std      elec    elec_std     total   total_std       vdw     vdw_std caprieval_rank\n",
        "1       2       3       yes     -119.238 6.928   1.383   0.087   0.816   0.035   3.729   0.278   0.732   0.013   2.949   0.135   1.468   0.087   129.217 13.524  1850.680        36.361  4.073   3.189   -468.883        30.770  -382.122        23.429  -42.456 8.236   1\n",
        "2       1       4       -        -97.753 7.662   5.037   0.124   0.134   0.007   12.077  0.969   0.183   0.014   10.874  0.328   4.494   0.261   166.356 17.690  1585.855        62.659  5.596   2.998   -272.086        38.750  -171.297        33.845  -65.568 8.237   2\n",
        "3       3       2       yes      -84.236 8.511   10.756  0.411   0.069   0.017   20.253  0.596   0.079   0.003   18.837  0.562   11.044  0.347   136.880 11.285  1397.195        38.715  5.962   0.830   -327.803        47.780  -229.249        56.062  -38.325 3.004   3\n",
        "```\n",
        "\n",
        "\n",
        "In this file you find the cluster rank (which corresponds to the naming of the clusters in the previous `seletop` directory), the cluster ID (which is related to the size of the cluster, 1 being always the largest cluster), the number of models (n) in the cluster and the corresponding statistics (averages + standard deviations). The corresponding cluster PDB files will be found in the preceeding `09_seletopclusts` directory.\n",
        "\n",
        "While these simple text files can be easily checked from the command line already, they might be cumbersome to read.\n",
        "For that reason, we have developed a post-processing analysis that automatically generates html reports for all `caprieval` steps in the workflow.\n",
        "These are located in the respective `analysis/XX_caprieval` directories and can be viewed using your favorite web browser.\n",
        "\n"
      ],
      "metadata": {
        "id": "8xvUhJanzKHn"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "---\n",
        "\n",
        "## Cluster statistics\n",
        "\n",
        "Let us now analyse the docking results. Use for that either your own run or a pre-calculated run provided in the `runs` directory.\n",
        "The `analysis/10_caprieval_analysis` directory of the respective run directory contains various html files. To visualise the report directly in your browser, download the files to your computer and open those in a web browser.  The `report.html` contains interactive plots that may take some time to generate.\n",
        "\n",
        "On the top of the page, you will see a table that summarises the cluster statistics (taken from the `capri_clt.tsv` file).\n",
        "The columns (corresponding to the various clusters) are sorted by default on the cluster rank, which is based on the HADDOCK score.\n"
      ],
      "metadata": {
        "id": "_AwXwt8xz19M"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title View the cluster statistics table\n",
        "import json\n",
        "import re\n",
        "from plotly.offline import iplot\n",
        "from IPython.display import display, HTML\n",
        "import pandas as pd\n",
        "\n",
        "# Specify the path to your HTML file\n",
        "html_file_path = PROJECT_DIR / \"runs\" / \"run1\" / \"analysis\" / \"10_caprieval_analysis/report.html\"\n",
        "\n",
        "# Read the HTML file\n",
        "with open(html_file_path, 'r') as f:\n",
        "    html_content = f.read()\n",
        "\n",
        "# Use a regular expression to find the script tag with id=\"datatable2\" and extract its content\n",
        "match_table = re.search(r'<script id=\"datatable2\" type=\"application/json\">\\s*(.*?)\\s*</script>', html_content, re.DOTALL)\n",
        "\n",
        "if match_table:\n",
        "    json_data_str_table = match_table.group(1)\n",
        "    # Load the JSON data\n",
        "    table_data = json.loads(json_data_str_table)\n",
        "\n",
        "    # Extract the 'clusters' list from the JSON data\n",
        "    clusters_data = table_data.get('clusters', [])\n",
        "\n",
        "    # Create a pandas DataFrame from the clusters data\n",
        "    df = pd.json_normalize(clusters_data)\n",
        "\n",
        "    # Drop the columns related to best cluster files\n",
        "    cols_to_drop = ['under_eval', 'total', 'rmsd', 'best1', 'best2', 'best3', 'best4', 'ilrmsd.mean', 'ilrmsd.std', 'air.mean', 'air.std', 'desolv.mean', 'desolv.std', 'elec.mean', 'elec.std', 'vdw.mean', 'vdw.std'] # Add more if there are more 'best' columns\n",
        "    df = df.drop(columns=[col for col in cols_to_drop if col in df.columns])\n",
        "\n",
        "    # Display the DataFrame\n",
        "    display(df)\n",
        "else:\n",
        "    print(\"Could not find the script tag with id='datatable2' in the HTML file.\")\n"
      ],
      "metadata": {
        "id": "aVhH_-cg0Orj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Inspect the final cluster statistics.\n",
        "\n",
        "__*How many clusters have been generated?*__\n",
        "\n",
        "__*Look at the score of the first few clusters: Are they significantly different if you consider their average scores and standard deviations?*__\n",
        "\n",
        "Since for this tutorial we have at hand the crystal structure of the complex, we provided it as reference to the `caprieval` modules.\n",
        "This means that the iRMSD, lRMSD, Fnat and DockQ statistics report on the quality of the docked model compared to the reference crystal structure.\n",
        "\n",
        "CAPRI is using the following criteria to assess the quality of a model:\n",
        "\n",
        "* **Unacceptable**:    i-RMSD >4Å or Fnat < 0.1\n",
        "* **Acceptable**:      4Å <=i-RMSD < 2Å and Fnat > 0.1\n",
        "* **Medium**:          2Å <=i-RMSD < 1Å and Fnat > 0.3\n",
        "* **High**:            i-RMSD < 1Å and Fnat > 0.5\n",
        "\n",
        "__*How many clusters of acceptable or better quality have been generate according to CAPRI criteria?*__\n",
        "\n",
        "__*What is the rank of the best cluster generated?*__\n",
        "\n",
        "__*What is the cluster ID of the best cluster generated?*__\n",
        "\n",
        "The cluster ID is related to the size of the cluster with cluster 1 being the largest one.\n",
        "\n",
        "__*What is the rank of the first acceptable of better cluster generated?*__"
      ],
      "metadata": {
        "id": "saX02tqI0PW1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Visualizing the scores and their components\n",
        "\n",
        "\n",
        "Next to the cluster statistic table shown above, the `report.html` file also contains a variety of plots displaying the HADDOCK score\n",
        "and its components against various CAPRI metrics (i-RMSD, l-RMSD,  Fnat, Dock-Q) with a color-coded representation of the clusters.\n",
        "These are interactive plots. A menu on the top right of the first row (you might have to scroll to the rigth to see it)\n",
        "allows you to zoom in and out in the plots and turn on and off clusters.\n",
        "\n",
        "To view theses plots in the notebook execute the following cell.\n",
        "\n",
        "__*Examine the plots (remember here that higher DockQ values and lower i-RMSD values correspond to better models)*__\n"
      ],
      "metadata": {
        "id": "eSz-FyQD-e7k"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title View the various statistics plots\n",
        "import json\n",
        "import re\n",
        "from plotly.offline import iplot\n",
        "from IPython.display import display, HTML\n",
        "import pandas as pd\n",
        "\n",
        "# Specify the path to your HTML file\n",
        "html_file_path = PROJECT_DIR  / \"runs\" / \"run1\" / \"analysis\" / \"10_caprieval_analysis/report.html\"\n",
        "\n",
        "# Read the HTML file\n",
        "with open(html_file_path, 'r') as f:\n",
        "    html_content = f.read()\n",
        "\n",
        "# Use a regular expression to find the script tag with id=\"data1\" and extract its content\n",
        "match_plot = re.search(r'<script id=\"data1\" type=\"application/json\">\\s*(.*?)\\s*</script>', html_content, re.DOTALL)\n",
        "\n",
        "if match_plot:\n",
        "    json_data_str_plot = match_plot.group(1)\n",
        "    # Load the JSON data\n",
        "    plot_data = json.loads(json_data_str_plot)\n",
        "\n",
        "    # Set width and height to None to make the plot responsive\n",
        "    if 'layout' in plot_data:\n",
        "        plot_data['layout']['width'] = None\n",
        "        #plot_data['layout']['height'] = None\n",
        "\n",
        "    # Display the plot using iplot\n",
        "    iplot(plot_data)\n",
        "else:\n",
        "    print(\"Could not find the script tag with id='data1' in the HTML file.\")\n"
      ],
      "metadata": {
        "id": "8e5nwl5V-foB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Finally, the report also shows plots of the cluster statistics (distributions of values per cluster ordered according to their HADDOCK rank).\n",
        "\n",
        "To view those distribution plots execute the next cell.\n",
        "\n",
        "__*For this antibody-antigen case, which of the score components correlates best with the quality of the models?*__\n",
        "\n"
      ],
      "metadata": {
        "id": "Vqltay5iDasI"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title View the distribution plots\n",
        "import json\n",
        "import re\n",
        "from plotly.offline import iplot\n",
        "from IPython.display import display, HTML\n",
        "import pandas as pd\n",
        "\n",
        "# Specify the path to your HTML file\n",
        "html_file_path = PROJECT_DIR / \"runs\" / \"run1\" / \"analysis\" / \"10_caprieval_analysis/report.html\"\n",
        "\n",
        "# Read the HTML file\n",
        "with open(html_file_path, 'r') as f:\n",
        "    html_content = f.read()\n",
        "\n",
        "# Use a regular expression to find the script tag with id=\"data1\" and extract its content\n",
        "match_plot = re.search(r'<script id=\"data2\" type=\"application/json\">\\s*(.*?)\\s*</script>', html_content, re.DOTALL)\n",
        "\n",
        "if match_plot:\n",
        "    json_data_str_plot = match_plot.group(1)\n",
        "    # Load the JSON data\n",
        "    plot_data = json.loads(json_data_str_plot)\n",
        "    # Set width and height to None to make the plot responsive\n",
        "    if 'layout' in plot_data:\n",
        "        plot_data['layout']['width'] = None\n",
        "        #plot_data['layout']['height'] = None\n",
        "\n",
        "    # Display the plot using iplot\n",
        "    iplot(plot_data)\n",
        "else:\n",
        "    print(\"Could not find the script tag with id='data1' in the HTML file.\")"
      ],
      "metadata": {
        "id": "3sUx6zJK_KP3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "## Some single structure analysis\n",
        "\n",
        "\n",
        "Going back to command line analysis, we are providing in the `scripts` directory a simple script that extracts some model statistics for acceptable or better models from the `caprieval` steps.\n",
        "To use it, simply call the script with as argument the run directory you want to analyze, e.g.:\n",
        "\n",
        "```\n",
        "scripts/extract-capri-stats.sh runs/run1\n",
        "```\n",
        "\n",
        "_**Note**_ that this kind of analysis only makes sense when we know the reference complex and for benchmarking / performance analysis purposes.\n",
        "\n"
      ],
      "metadata": {
        "id": "GayTZXpuABM_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!scripts/extract-capri-stats.sh runs/run1"
      ],
      "metadata": {
        "id": "nvqoySu07NWG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Look at the single structure statistics provided by the script.\n",
        "\n",
        "__*How does the quality of the best model changes after flexible refinement? Consider here the various metrics.*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <i>Answer</i> <i class=\"material-icons\">expand_more</i>\n",
        "  </summary>\n",
        "  <p>\n",
        "  In terms of iRMSD values, we only observe very small differences in the best model.\n",
        "  The fraction of native contacts and the DockQ scores are however improving much more after flexible refinement but increases again slightly after final minimisation.\n",
        "  All this will of course depend on how different are the bound and unbound conformations and the amount of data used to drive the docking process.\n",
        "  In general, from our experience, the more and better data at hand, the larger the conformational changes that can be induced.\n",
        "  </p>\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "__*Is the best model always ranked first?*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <i>Answer</i> <i class=\"material-icons\">expand_more</i>\n",
        "  </summary>\n",
        "  <p>\n",
        "    This is not the case. The scoring function is not perfect, but does a reasonable job at ranking models of acceptable or better quality on top in this case.\n",
        "  </p>\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "**_Note_**: A similar script to extract cluster statistics is available in the `scripts` directory as `extract-capri-stats-clt.sh`.\n",
        "\n"
      ],
      "metadata": {
        "id": "9F5M54607Ob_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "## Contacts analysis\n",
        "\n",
        "We have added a contact analysis module to HADDOCK3 that generates for each cluster both a contact matrix of the entire system showing all contacts within a 4.5Å cutoff and a chord chart representation of intermolecular contacts.\n",
        "\n",
        "In the current workflow we run, those files can be found in the `11_contactmap` directory. These are again html files with interactive plots (hover with your mouse over the plots).\n",
        "\n",
        "This file taken from the pre-computed run can be visualized by executing the next cell which will visualize the contacts of the first ranked cluster (cluster 2).\n",
        "\n",
        "__*Can you identify which residue(s) make(s) the most intermolecular contacts?*__\n",
        "\n"
      ],
      "metadata": {
        "id": "DbcuwvW-Ag4o"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title View the contact chord chart\n",
        "import json\n",
        "import re\n",
        "from plotly.offline import iplot\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "# Specify the path to the HTML file containing the chord chart\n",
        "html_file_path = PROJECT_DIR / \"runs\" / \"run1\" / \"11_contactmap\" / \"cluster2_rank1_chordchart.html\"\n",
        "\n",
        "# Read the HTML file\n",
        "with open(html_file_path, 'r') as f:\n",
        "    html_content = f.read()\n",
        "\n",
        "# Use a regular expression to find the script tag containing the plot data\n",
        "# The id of the script tag might vary, so a more general pattern is used\n",
        "# We assume the plot data is within a script tag with type=\"application/json\"\n",
        "match_plot = re.search(r'<script.*?type=\"application/json\".*?>\\s*(.*?)\\s*</script>', html_content, re.DOTALL)\n",
        "\n",
        "if match_plot:\n",
        "    json_data_str_plot = match_plot.group(1)\n",
        "    # Load the JSON data\n",
        "    plot_data = json.loads(json_data_str_plot)\n",
        "    # Display the plot using iplot\n",
        "    iplot(plot_data)\n",
        "else:\n",
        "    print(f\"Could not find the plot data in the HTML file at {html_file_path}\")"
      ],
      "metadata": {
        "id": "e4KiyiN2A5Qx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "## Visualization of the models\n",
        "\n",
        "We can visualize the models from the various clusters by execting the next cell (edit it to change the model you want to visualize). The model visualised here is the top ranked model of the top ranker cluster (cluster 2).\n",
        "\n",
        "For more detailed visualisation using PyMOL (after downloading the results to your local computer refer to our [online tutorial](https://www.bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/#visualization-of-the-models).\n",
        "\n"
      ],
      "metadata": {
        "id": "OvqCZa7pPNxd"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Let’s now check if the active residues which we have defined (the paratope and epitope) are actually part of the interface. For this execute the next cell.\n",
        "\n",
        "The active residues of the antibody (paratope) are shown in red, those of the antigen (epitope) in orange and the passive residues of the antigen in green.\n",
        "\n",
        "__*Are the residues of the paratope and NMR epitope at the interface?*__\n",
        "\n"
      ],
      "metadata": {
        "id": "0JCDFqWtPRzm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualize the active and passive residues in the model\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"runs\" / \"run1\" / \"09_seletopclusts\" / \"cluster_2_model_1.pdb.gz\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {'color': 'white'}})\n",
        "    view.addStyle({'resi': [31,32,33,34,35,52,54,55,56,100,101,102,103,104,105,106,151,152,169,170,173,211,212,213,214,216], 'chain': 'A'},{'cartoon':{'color':'red'}})\n",
        "    view.addStyle({'resi': [31,32,33,34,35,52,54,55,56,100,101,102,103,104,105,106,151,152,169,170,173,211,212,213,214,216], 'chain': 'A'},{'stick':{'color':'red'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117], 'chain': 'B'},{'cartoon':{'color':'organge'}})\n",
        "    view.addStyle({'resi': [72,73,74,75,81,83,84,89,90,92,94,96,97,98,115,116,117], 'chain': 'B'},{'cartoon':{'color':'orange'}})\n",
        "    view.addStyle({'resi': [3,24,46,47,48,50,66,76,77,79,80,82,86,87,88,91,93,95,118,119,120], 'chain': 'B'},{'cartoon':{'color':'green'}})\n",
        "    view.addStyle({'resi': [3,24,46,47,48,50,66,76,77,79,80,82,86,87,88,91,93,95,118,119,120], 'chain': 'B'},{'stick':{'color':'green'}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "_9y3FjW5IZ4G"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title View the top ranking model from the best cluster\n",
        "#@markdown It will be shown superimposed onto the reference structure\n",
        "# Define paths to PDB files to be aligned\n",
        "# Here we are overlaying the reference and 1st model from the 1st cluster\n",
        "path1 = PROJECT_DIR / \"pdbs\" / \"4G6M_matched.pdb\"\n",
        "path2 = PROJECT_DIR / \"runs\" / \"run1\" / \"09_seletopclusts\" / \"cluster_1_model_1.pdb.gz\"\n",
        "\n",
        "# Align molecules by chain and display the result.\n",
        "# For the full alignement put all chain IDs in the `chains` variable.\n",
        "align_full(str(path1), str(path2), chains = ['A','B'])"
      ],
      "metadata": {
        "id": "lYMWxm9R6pU5",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "In our [online tutorial](https://www.bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/#visualization-of-the-models) you can find instructions to compare the models with the reference structure using PyMOL.\n"
      ],
      "metadata": {
        "id": "pHroy0YPIakp"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "# Conclusions\n",
        "\n",
        "We have demonstrated the usage of HADDOCK3 in an antibody-antigen docking scenario making use of the paratope information on the antibody side (i.e. no prior experimental information, but computational predictions) and an NMR-mapped epitope for the antigen.\n",
        "Compared to the static HADDOCK2.X workflow, the modularity and flexibility of HADDOCK3 allow to customise the docking protocols and to run a deeper analysis of the results.\n",
        "HADDOCK3's intrinsic flexibility can be used to improve the performance of antibody-antigen modelling compared to the results we presented in our\n",
        "[Structure 2020](https://doi.org/10.1016/j.str.2019.10.011) article and in the [related HADDOCK2.4 tutorial](/education/HADDOCK24/HADDOCK24-antibody-antigen).\n",
        "\n",
        "<br>\n",
        "<br>\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "wtUDO1fLKa2h"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# Congratulations! 🎉\n",
        "\n",
        "You have completed this tutorial. If you have any questions or suggestions, feel free to contact us via email or asking a question through our [support center](https://ask.bioexcel.eu).\n",
        "\n",
        "And check also our [education web page](https://bonvinlab.org/education) where you will find more tutorials!\n",
        "\n",
        "But there is more: Check the BONUSES!\n"
      ],
      "metadata": {
        "id": "zZuBwzb1lepS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "# BONUS1: Dissecting the interface energetics: what is the impact of a single mutation?\n",
        "\n",
        "Mutations at the binding interfaces can have widely varying effects on binding affinity - some may be negligible, while others can significantly strengthen or weaken the interaction. Exploring these mutations helps identify critical amino acids for redesigning structurally characterized protein-protein interfaces, which paves the way for developing protein-based therapeutics to deal with a diverse range of diseases.\n",
        "To pinpoint such amino acids positions, the residues across the protein interaction surfaces are either randomly or strategically mutated. Scanning mutations in this manner is experimentally costly. Therefore, computational methods have been developed to estimate the impact of an interfacial mutation on protein-protein interactions.\n",
        "These computational methods come in two main flavours. One involves rigorous free energy calculations, and, while highly accurate, these methods are computationally expensive. The other category includes faster, approximate approaches that predict changes in binding energy using statistical potentials, machine learning, empirical scoring functions etc. Though less precise, these faster methods are practical for large-scale screening and early-stage analysis. In this bonus exercise, we will take a look at two quick ways of estimating the effect of a single mutation in the interface.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "ukep8dflNAUw"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "## PROT-ON and haddock3-scoring to inspect a single mutation\n",
        "\n",
        "PROT-ON (Structure-based detection of designer mutations in PROTein-protein interface mutatiONs) is a tool and [online server](http://proton.tools.ibg.edu.tr:8001/about) that scans all possible interfacial mutations and **predicts ΔΔG score** by using EvoEF1 (active in both on the web server and stand-alone versions) or FoldX (active only in the stand-alone version) with the aim of finding the most mutable positions. The original publication describing PROT-ON can be found [here](https://www.frontiersin.org/journals/molecular-biosciences/articles/10.3389/fmolb.2023.1063971/full).\n",
        "\n",
        "Here we will use PROT-ON to analyse the interface of our antibody-antigen complex. For that, we will use the provided matched reference structure (`4G6M-matched.pdb`) in which both chains of the antibody have the same chainID (A), which allows us to analyse all interface residues of the antibody at once.\n",
        "\n",
        "__Note:__ Pre-calculated PROT-ON results for this system can be accessed [here](http://proton.tools.ibg.edu.tr:8001/result/ebcdec31308c46acb82e8010f7f21df1).\n",
        "\n",
        "\n",
        "Connect to the PROT-ON server page (link above) and fill in the following fields:\n",
        "\n",
        "\n",
        "```\n",
        "Specify your run name --> 4G6M_matched\n",
        "```\n",
        "\n",
        "```\n",
        "Choose / Upload your protein complex --> Select the provided _4G6M-matched.pdb_ file\n",
        "```\n",
        "\n",
        "```\n",
        "Which dimer chains should be analyzed --> Select chain A for the 1st molecule and B for the 2nd\n",
        "```\n",
        "\n",
        "```\n",
        "Pick the monomer for mutational scanning --> Select the first molecule - the antibody (toggle the switch ON under the chain A)\n",
        "```\n",
        "\n",
        "```\n",
        "Click on the Submit button\n",
        "```\n",
        "\n",
        "Your run should complete in 5-10 minutes. Once finished, you will be presented with a result page summarising the most depleting (ones that decrease the binding affinity) and most enriching (ones that increase the binding affinity) mutations.\n",
        "\n",
        "__*Which possible mutation would you propose to improve the binding affinity of the antibody?*__\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>See answer</i></b> <i class=\"material-icons\">expand_more</i>\n",
        "  </summary>\n",
        "The most enriching mutation is S150W with a -3.69 ΔΔG score.\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "Visualise the propose mutated residue in the 3D model (for this run the next cell). The mutated residue (change the residue selection if needed) will be displayed as spheres.\n",
        "\n",
        "__*Can you rationalise why it might increase the affinity?*__\n"
      ],
      "metadata": {
        "id": "PHEkdf4GNS1z"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualise the most enriching (increasing the interaction energy) mutation\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"pdbs\" / \"4G6M_matched.pdb\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {}})\n",
        "    view.addStyle({'resi': '150', 'chain': 'A'},{'sphere':{}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "tFiFimX0OIQ7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "With HADDOCK3, it is possible to take a step further. To perform the mutation, simply rename the desired residue and score such model - HADDOCK will take care of the topology regardless of the side chain differences and energy minimisation of the model. To do so, change the name of the residue to be mutated in the _4G6M-matched.pdb_ PDB file and save it to a new file. This can be done in the terminal with the following command:\n",
        "\n",
        "```\n",
        "sed 's/SER\\ A\\ 150/TRP\\ A\\ 150/g' pdbs/4G6M_matched.pdb > 4G6M_matched_S150W.pdb\n",
        "```\n",
        "\n",
        "Next, score the mutant using the command-line tool `haddock3-score`.\n",
        "This tool performs a short workflow composed of the `topoaa` and `emscoring` modules. Use flag `--outputpdb` to save energy-minimized model:\n",
        "\n",
        "```\n",
        "haddock3-score 4G6M_matched_S150W.pdb --outputpdb\n",
        "```\n",
        "\n",
        "Use _haddock3-score_ to calculate the score of the wild type, 4G6M-matched.pdb, and the mutation, 4G6M_matched_S150W.pdb.\n",
        "\n"
      ],
      "metadata": {
        "id": "q_coziwKN4m9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sed 's/SER\\ A\\ 150/TRP\\ A\\ 150/g' pdbs/4G6M_matched.pdb > 4G6M_matched_S150W.pdb\n",
        "!haddock3-score pdbs/4G6M_matched.pdb --outputpdb\n",
        "!haddock3-score 4G6M_matched_S150W.pdb --outputpdb"
      ],
      "metadata": {
        "id": "4F8HLZsm8Ng7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "__*Do you see a difference between wild-type and mutant scores?\n",
        "Might such single-residue mutation affect the binding affinity?*__\n",
        "\n",
        "\n",
        "Inspect the energy-minimized mutant model (4G6M_matched_S150W_hs.pdb) visually (for this run the next cell).\n",
        "\n",
        "__*Can you rationalise why such a mutation might increase the affinity?*__\n",
        "\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"bold\">\n",
        "    <b><i>Zoom in on the mutated residue</i></b>\n",
        "  </summary>\n",
        "  <figure style=\"text-align: center;\">\n",
        "   <img width=\"100%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/mutant-stacking.png\">\n",
        "   <center>\n",
        "   <i>TRP150 is stacking with TYR24 </i>\n",
        "   </center>\n",
        "  </figure>\n",
        "  <br>\n",
        "</details>\n",
        "<br>\n",
        "\n"
      ],
      "metadata": {
        "id": "CQ8QwN9B8Wbr"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualise the S150W mutation\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"pdbs\" / \"4G6M_matched_S150W_hs.pdb\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {}})\n",
        "    view.addStyle({'resi': '150', 'chain': 'A'},{'sphere':{}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()\n"
      ],
      "metadata": {
        "id": "dGBGkfpAQiY1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "## Alanine Scanning module\n",
        "\n",
        "Another way of exploring interface energetics is by using the `alascan` module of HADDOCK3. `alascan` stands for \"Alanine Scanning module\".\n",
        "\n",
        "This module is capable of mutating interface residues to Alanine and calculating the **Δ HADDOCK score** between the wild-type and mutant, thus providing a measure of the impact of each individual mutation. It is possible to scan all interface residues one by one or limit this scanning to a selected by user set of residues. By default, the mutation to Alanine is performed, as its side chain is just a methyl group, so side chain perturbations are minimal, as well as possible secondary structure changes. It is possible to perform the mutation to any other amino acid type - at your own risk, as such mutations may introduce structural uncertainty.\n",
        "\n",
        "**Important**: 1/ `alascan` calculates the difference between wild-type score vs mutant score, i.e. positive `Δscore` indicative of the enriched (stronger) binding and negative `Δscore` is indicative of the depleted (weaker) binding; 2/ Inside `alascan`, a short energy minimization of an input structure is performed, i.e. there's no need to include an additional refinement module prior to `alascan`.\n",
        "\n",
        "Here is an example of the workflow to scan interface energetics:\n",
        "\n",
        "```\n",
        "# ====================================================================\n",
        "# Scanning interface residues with haddock3\n",
        "# ====================================================================\n",
        "\n",
        "# directory in which the scoring will be done\n",
        "run_dir = \"run-energetics-alascan\"\n",
        "\n",
        "# compute mode\n",
        "mode = \"local\"\n",
        "ncores = 50\n",
        "\n",
        "# Post-processing to generate statistics and plots\n",
        "postprocess = true\n",
        "clean = true\n",
        "\n",
        "molecules =  [\n",
        "    \"pdbs/4G6M_matched.pdb\",\n",
        "    ]\n",
        "\n",
        "# ====================================================================\n",
        "# Parameters for each stage are defined below\n",
        "# ====================================================================\n",
        "[topoaa]\n",
        "\n",
        "[emref]\n",
        "\n",
        "[alascan]\n",
        "# mutate each interface residue to Alanine\n",
        "scan_residue = 'ALA'\n",
        "# generate plot of delta score and its components per each mutation\n",
        "plot = true\n",
        "\n",
        "# ====================================================================\n",
        "```\n",
        "\n",
        "\n",
        "An alanine scaning scenario configuration file is provided in the `workflows/` directory as `interaction-energetics.cfg`, and precomputed results are in `runs/run-energetics-alascan`.\n"
      ],
      "metadata": {
        "id": "23sQ4bf2Q4WG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Run the haddock3 alanine scanning workflow\n",
        "!haddock3 workflows/interaction-energetics.cfg"
      ],
      "metadata": {
        "id": "QU2VSBchsx0W",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "The output folder contains, among others, a directory titled `2_alascan` with a file `scan_clt_unclustered.tsv` that lists each mutation, corresponding score and individual terms:\n",
        "\n",
        "```\n",
        "################################################################################\n",
        "# `alascan` cluster results for cluster unclustered\n",
        "# reported values are the average for the cluster\n",
        "#\n",
        "# z_score is calculated with respect to the mean values of all residues\n",
        "################################################################################\n",
        "chain   resid   resname full_resname    delta_score     delta_score_std delta_vdw       delta_vdw_std   delta_elec      delta_elec_std  delta_desolv    delta_desolv_std        delta_bsa       delta_bsa_std   frac_pres       z_score\n",
        "A       32      SER     A-32-SER        4.71    0.00    1.49    0.00    5.66    0.00    2.08    0.00    24.02   0.00    1.00    1.18\n",
        "A       33      GLY     A-33-GLY        2.55    0.00    3.23    0.00    0.03    0.00    -0.69   0.00    16.31   0.00    1.00    0.85\n",
        "A       54      TRP     A-54-TRP        0.16    0.00    -2.11   0.00    9.29    0.00    0.41    0.00    12.26   0.00    1.00    0.49\n",
        "A       55      TRP     A-55-TRP        -5.45   0.00    -2.92   0.00    5.52    0.00    -3.64   0.00    59.08   0.00    1.00    -0.35\n",
        "A       56      ASP     A-56-ASP        -12.38  0.00    1.26    0.00    -81.09  0.00    2.58    0.00    12.53   0.00    1.00    -1.39\n",
        "A       58      ASP     A-58-ASP        -21.75  0.00    2.24    0.00    -135.74 0.00    3.16    0.00    39.68   0.00    1.00    -2.80\n",
        "A       59      GLU     A-59-GLU        0.19    0.00    0.11    0.00    -1.14   0.00    0.31    0.00    12.42   0.00    1.00    0.50\n",
        "A       60      SER     A-60-SER        2.58    0.00    3.12    0.00    -10.76  0.00    1.62    0.00    17.58   0.00    1.00    0.86\n",
        "...\n",
        "```\n",
        "\n",
        "__*Can you identify the most enriching/depleting mutation of each chain? *__\n",
        "\n",
        "You can use an additional script `scripts/get-alascan-extrema.sh` to check your answer:\n"
      ],
      "metadata": {
        "id": "4NZbZWfesyYp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!bash scripts/get-alascan-extrema.sh runs/run-energetics-alascan/2_alascan/scan_clt_unclustered.tsv"
      ],
      "metadata": {
        "id": "jmDSKwBS9DUB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Mutation of the residue ASP58 turned out to be the most depleting within chain A.\n",
        "\n",
        "Let us visualise the impact of each mutated residue. This can be done by looking at the interactive plot generated in the `2_alascan` directory called `scan_clt_unclustered.html`.\n",
        "\n",
        "In the generated interactive plot, negative values are indicative of mutations that should destabilize the complex.\n"
      ],
      "metadata": {
        "id": "YHq5hkiJ9Dsg"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualise the results of the alanine scanning\n",
        "import json\n",
        "import re\n",
        "from plotly.offline import iplot\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "# Specify the path to the HTML file\n",
        "html_file_path = PROJECT_DIR / \"runs\" / \"run-energetics-alascan\" / \"2_alascan\" / \"scan_clt_unclustered.html\"\n",
        "\n",
        "# Read the HTML file\n",
        "try:\n",
        "    with open(html_file_path, 'r') as f:\n",
        "        html_content = f.read()\n",
        "\n",
        "    # Use a regular expression to find a script tag with type=\"application/json\" and extract its content\n",
        "    # The id of the script tag might vary, so a general pattern is used\n",
        "    match_plot = re.search(r'<script id=\"data1\" type=\"application/json\">\\s*(.*?)\\s*</script>', html_content, re.DOTALL)\n",
        "\n",
        "    if match_plot:\n",
        "        json_data_str_plot = match_plot.group(1)\n",
        "        # Load the JSON data\n",
        "        plot_data = json.loads(json_data_str_plot)\n",
        "\n",
        "        # Set width and height to None to make the plot responsive\n",
        "        if 'layout' in plot_data:\n",
        "            plot_data['layout']['width'] = None\n",
        "            plot_data['layout']['height'] = None\n",
        "\n",
        "        # Display the plot using iplot\n",
        "        iplot(plot_data)\n",
        "    else:\n",
        "        print(\"Could not find JSON plot data in the HTML file.\")\n",
        "\n",
        "except FileNotFoundError:\n",
        "    print(f\"Error: File not found at {html_file_path}\")\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred: {e}\")"
      ],
      "metadata": {
        "id": "Vw5bSpKSU8OT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "__*Can you identify the most enriching/depleting mutation of each chain?*__\n",
        "\n",
        "You can use an additional script /scripts/get-alascan-extrema.sh to check your answer:\n"
      ],
      "metadata": {
        "id": "FOpu6shaU76Q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!bash scripts/get-alascan-extrema.sh runs/run-energetics-alascan/2_alascan/scan_clt_unclustered.tsv"
      ],
      "metadata": {
        "id": "H8tXvJCE-DSg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Mutation of the residue ASP58 turned out to be the most depleting within chain A.\n"
      ],
      "metadata": {
        "id": "0aqXCoDV-Dr8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualise the Asp58 environemnt\n",
        "import py3Dmol\n",
        "import gzip\n",
        "import os\n",
        "\n",
        "# Create a 3Dmol viewer object\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "\n",
        "# Specify the path to your PDB file (can be .pdb or .pdb.gz)\n",
        "pdb_file_path = PROJECT_DIR / \"pdbs\" / \"4G6M_matched_S150W_hs.pdb\"\n",
        "\n",
        "# Check if the file exists\n",
        "if not os.path.exists(pdb_file_path):\n",
        "    print(f\"Error: File not found at {pdb_file_path}\")\n",
        "else:\n",
        "    # Check if the file is gzipped\n",
        "    if str(pdb_file_path).endswith('.gz'):\n",
        "        with gzip.open(pdb_file_path, 'rt') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "    else:\n",
        "        with open(pdb_file_path, 'r') as f:\n",
        "            view.addModel(f.read(), 'pdb')\n",
        "\n",
        "    view.setStyle({'cartoon': {'color': 'white'}})\n",
        "    view.addStyle({'stick': {}})\n",
        "    view.addStyle({'resi': '58', 'chain': 'A'},{'sphere':{}})\n",
        "    view.addSurface(py3Dmol.SAS, {'opacity':0.4, 'color':'white'})\n",
        "    view.zoomTo()\n",
        "    view.show()"
      ],
      "metadata": {
        "id": "JEFIavyLSbxG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You should be able to see that ASP58 is close to two lysines: LYS98, and LYS94.\n",
        "\n",
        "<figure style=\"text-align: center;\">\n",
        "  <img width=\"50%\" src=\"https://bonvinlab.org/education/HADDOCK3/HADDOCK3-antibody-antigen/asp58_contacts.png\">\n",
        "  <center>\n",
        "  <i>ASP58 makes h-bonds with two neighbouring residues</i>\n",
        "  </center>\n",
        "</figure>\n",
        "\n"
      ],
      "metadata": {
        "id": "MOrt_ZTmScIv"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "---\n",
        "\n",
        "# BONUS 2: Running a scoring scenario\n",
        "\n",
        "This section demonstrates the use of HADDOCK3 to score the various models obtained using HADDOCK and AlphaFold predictions and observe if the HADDOCK scoring function is able to detect the quality of the models.\n",
        "\n",
        "To this end the following workflow is defined:\n",
        "\n",
        "1. Generate the topologies for the various models.\n",
        "2. Energy Minimise all complexes.\n",
        "3. Cluster the models using Fraction of Common Contacts:\n",
        "  - set the parameter `min_population` to 1 so that all models, including singletons (models that do not cluster with any others), will be forwarded to the next steps.\n",
        "  - set the parameter `plot_matrix` to true to generate a matrix of the clusters for a visual representation.\n",
        "4. Comparison of the models with the reference complex `4G6M_matched.pdb` using CAPRI criteria.\n",
        "\n",
        "For this, two ensembles must be scored and one structure will be used as a reference. You can find them in the `pdbs/` directory:\n",
        "- `07_emref_and_top5af2_ensemble.pdb`: An ensemble of models obtained from the ensemble run, combined with top5 AlphaFold2 predictions.\n",
        "- `af3server_15052024_top5ens.pdb`: An ensemble of top5 AlphaFold3 predictions.\n",
        "- `4G6M_matched.pdb`: The reference structure for quality assessments.\n",
        "\n",
        "\n",
        "```\n",
        "# ====================================================================\n",
        "# Antibody-antigen docking example with restraints from the antibody\n",
        "# paratope to the NMR-identified epitope on the antigen\n",
        "# ====================================================================\n",
        "run_dir = \"scoring-haddock3-alphafold2and3-ensemble\"\n",
        "\n",
        "molecules =  [\n",
        "    \"pdbs/haddock3-ens-emref-ensemble.pdb\",\n",
        "    \"pdbs/af2-models.pdb\",\n",
        "    \"pdbs/af3-models.pdb\",\n",
        "    ]\n",
        "\n",
        "# ====================================================================\n",
        "# Parameters for each stage are defined below\n",
        "# ====================================================================\n",
        "\n",
        "[topoaa]\n",
        "\n",
        "[emscoring]\n",
        "\n",
        "[clustfcc]\n",
        "# Reduce the min_population to define a cluster to 1 so that models\n",
        "# that do not cluster with any other will define singlotons\n",
        "min_population = 1\n",
        "# Generate a matrix of the clusters\n",
        "plot_matrix = true\n",
        "\n",
        "[caprieval]\n",
        "reference_fname = \"4G6M_matched.pdb\"\n",
        "\n",
        "# ====================================================================\n",
        "```\n",
        "\n",
        "\n",
        "A scoring scenario configuration file is provided in the `workflows/` directory as `scoring-antibody-antigen.cfg, precomputed results in `runs/run-scoring`.\n"
      ],
      "metadata": {
        "id": "SVrBkoOEls66"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Run the haddock3 scoring workflow\n",
        "!haddock3 workflows/scoring-antibody-antigen.cfg"
      ],
      "metadata": {
        "id": "4cbhUaoIuZQ3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "You can look at the `capri_ss.tsv` file in the `4_caprieval` directory. It contains the energy minimised statistics:\n",
        "\n",
        "```\n",
        "model   md5     caprieval_rank  score   irmsd   fnat    lrmsd   ilrmsd  dockq   rmsd    cluster_id      cluster_ranking model-cluster_ranking   air     angles  bonds   bsa     cdih    coup    dani    desolv  dihe    elec    improper        rdcs    rg      sym     total   vdw     vean    xpcs\n",
        "../1_emscoring/emscoring_82.pdb -       1       -157.149        0.910   0.897   2.201   1.456   0.855   1.016   2       1       1       0.000   296.229 64.664  2000.130        0.000   0.000   0.000   7.345   2068.230        -599.183        84.513  0.000   0.000   0.000   -643.841        -44.658      0.000   0.000\n",
        "../1_emscoring/emscoring_2.pdb  -       2       -156.452        0.880   0.948   1.949   1.355   0.881   0.989   2       1       2       0.000   298.570 59.587  1914.860        0.000   0.000   0.000   3.125   2052.880        -504.372        85.779  0.000   0.000   0.000   -563.075        -58.703      0.000   0.000\n",
        "../1_emscoring/emscoring_64.pdb -       3       -138.214        1.052   0.914   3.039   1.955   0.824   1.294   2       1       3       0.000   286.361 53.633  1784.350        0.000   0.000   0.000   -2.359  2040.230        -424.542        77.017  0.000   0.000   0.000   -475.489        -50.947      0.000   0.000\n",
        "../1_emscoring/emscoring_40.pdb -       4       -135.230        1.085   0.897   1.866   1.756   0.836   1.144   2       1       4       0.000   299.170 56.868  1875.210        0.000   0.000   0.000   3.490   2034.370        -429.067        87.524  0.000   0.000   0.000   -481.973        -52.906      0.000   0.000\n",
        "../1_emscoring/emscoring_37.pdb -       5       -134.569        13.624  0.069   22.589  21.764  0.068   13.881  5       2       1       0.000   305.568 67.872  1802.890        0.000   0.000   0.000   6.081   2050.950        -426.815        93.743  0.000   0.000   0.000   -482.102        -55.287      0.000   0.000\n",
        "../1_emscoring/emscoring_22.pdb -       6       -129.571        4.855   0.155   11.612  10.609  0.197   4.234   8       3       1       0.000   309.609 62.278  1591.470        0.000   0.000   0.000   8.217   2058.920        -345.081        86.168  0.000   0.000   0.000   -413.853        -68.772      0.000   0.000\n",
        "...\n",
        "```\n",
        "\n",
        "__*\n",
        "Did the HADDOCK scoring do a good job at putting the best models on top (consider for example the DockQ score)?*__\n",
        "\n",
        "The `emscoring` module renames all models, which makes it difficult to know what was the original model.\n",
        "You can however trace back a model to its original file by looking into the `traceback/traceback.tsv` file:\n",
        "\n",
        "```\n",
        "00_topo1        1_emscoring     1_emscoring_rank\n",
        "emref_9_from_haddock3-ens-emref-ensemble_83_haddock.psf   emscoring_82.pdb        1\n",
        "emref_10_from_haddock3-ens-emref-ensemble_2_haddock.psf   emscoring_2.pdb        2\n",
        "emref_7_from_haddock3-ens-emref-ensemble_67_haddock.psf   emscoring_64.pdb        3\n",
        "emref_5_from_haddock3-ens-emref-ensemble_45_haddock.psf   emscoring_40.pdb        4\n",
        "emref_47_from_haddock3-ens-emref-ensemble_42_haddock.psf  emscoring_37.pdb        5\n",
        "emref_35_from_haddock3-ens-emref-ensemble_29_haddock.psf  emscoring_22.pdb        6\n",
        "...\n",
        "```\n",
        "\n",
        "__*Try to locate the AlphaFold2 and AlphaFold3 models (their filenames start with _abag_test_ and _af3server_, respectively)*__\n",
        "\n",
        "A simple way to extra this information is to use `grep`:\n"
      ],
      "metadata": {
        "id": "e5iQIYbruXUU"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep abag runs/run-scoring/traceback/traceback.tsv"
      ],
      "metadata": {
        "id": "gHuq65pK-yJD"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"font-weight: bold\">\n",
        "    <i>See the output of grep for the AlphaFold2 models</i>\n",
        "  <br>\n",
        "  </summary>\n",
        "  <pre>\n",
        "> grep abag traceback.tsv\n",
        "abagtest_2d03e_unrelaxed_rank_001_alphafold2_multimer_v3_model_3_seed_000_from_af2-models_1_haddock.psf\temscoring_84.pdb\t86\n",
        "abagtest_2d03e_unrelaxed_rank_005_alphafold2_multimer_v3_model_2_seed_000_from_af2-models_5_haddock.psf\temscoring_88.pdb\t90\n",
        "abagtest_2d03e_unrelaxed_rank_004_alphafold2_multimer_v3_model_4_seed_000_from_af2-models_4_haddock.psf\temscoring_87.pdb\t91\n",
        "abagtest_2d03e_unrelaxed_rank_003_alphafold2_multimer_v3_model_1_seed_000_from_af2-models_3_haddock.psf\temscoring_86.pdb\t92\n",
        "abagtest_2d03e_unrelaxed_rank_002_alphafold2_multimer_v3_model_5_seed_000_from_af2-models_2_haddock.psf\temscoring_85.pdb\t93\n",
        "  </pre>\n",
        "</details>\n",
        "<br>\n"
      ],
      "metadata": {
        "id": "Ymzg9VDd-yuW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep af3server runs/run-scoring/traceback/traceback.tsv"
      ],
      "metadata": {
        "id": "iiXWOVzF-zsA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"font-weight: bold\">\n",
        "    <i>See the output of grep for the AlphaFold3 models</i>\n",
        "  <br>\n",
        "  </summary>\n",
        "  <pre>\n",
        "> grep abag traceback.tsv\n",
        "af3server_15052024_2_ready_from_af3-models_2_haddock.psf\temscoring_90.pdb\t40\n",
        "af3server_15052024_1_ready_from_af3-models_1_haddock.psf\temscoring_89.pdb\t81\n",
        "af3server_15052024_4_ready_from_af3-models_4_haddock.psf\temscoring_92.pdb\t87\n",
        "af3server_15052024_3_ready_from_af3-models_3_haddock.psf\temscoring_91.pdb\t88\n",
        "af3server_15052024_5_ready_from_af3-models_5_haddock.psf\temscoring_93.pdb\t89\n",
        "  </pre>\n",
        "</details>\n",
        "<br>\n",
        "\n",
        "__*What are their ranks?*__\n",
        "\n",
        "We have already seen in the previous section that none of the AlphaFold models were close to the real complex. This is however also the case for some of the HADDOCK models. Still the AlpaFold models score very badly, toward the end of the ranked list of models.\n",
        "\n",
        "__*Having found their ranks, can you figure out from the statistics in _capri_ss.tsv_ which component of the HADDOCK score causes in particular this bad scoring?*__\n",
        "\n",
        "\n",
        "<details style=\"background-color:#DAE4E7\">\n",
        "  <summary style=\"font-weight: bold\">\n",
        "    <i>See the answer</i>\n",
        "  <br>\n",
        "  </summary>\n",
        "  <p> The bottom eight models (the worst ranking ones) are all AlphaFold3/2 models. Looking at the componenents of the score\n",
        "  (some were left out in the table below for simplicity) one can see that it is mainly the van der Waals energy that causes the high scores,\n",
        "  which is indicative of clashes in the models.</p>\n",
        "  <pre>\n",
        "model               md5 caprieval_rank  score    irmsd   fnat    lrmsd   ilrmsd  dockq   rmsd    bsa        desolv    elec      vdw vean    xpcs\n",
        "...\n",
        "../1_emscoring/emscoring_84.pdb -   86  -67.914  11.123  0.000   22.413  18.626  0.048   12.213  3535.520   -67.537  -150.913    29.806\n",
        "../1_emscoring/emscoring_92.pdb -   87  -63.263  11.426  0.000   22.104  21.035  0.049   11.048  1383.920    -9.924   -88.656   -35.607\n",
        "../1_emscoring/emscoring_91.pdb -   88  -50.990  13.665  0.000   23.793  22.150  0.042   13.796  1492.150    -8.962  -167.236    -8.581\n",
        "../1_emscoring/emscoring_93.pdb -   89  -46.871   6.644  0.000   10.617  11.333  0.146   6.455   1740.990    -8.906   -35.623   -30.841\n",
        "../1_emscoring/emscoring_88.pdb -   90   48.283  12.919  0.000   20.484  19.885  0.053   14.706  3914.250   -68.786  -129.461   142.961\n",
        "../1_emscoring/emscoring_87.pdb -   91  180.468  12.447  0.000   22.153  19.299  0.048   14.160  3639.430   -66.857  -240.130   295.351\n",
        "../1_emscoring/emscoring_86.pdb -   92  240.307  12.572  0.000   21.662  19.799  0.049   14.187  3535.820   -69.380  -154.703   340.628\n",
        "../1_emscoring/emscoring_85.pdb -   93  781.210  15.174  0.000   23.497  24.993  0.042   17.151  3278.340   -61.261   -86.026   859.677\n",
        "  </pre>\n",
        "</details>\n",
        "<br>\n"
      ],
      "metadata": {
        "id": "UI3Gs3Pd-0CL"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "# Congratulations! 🎉\n",
        "\n",
        "You have completed this tutorial. If you have any questions or suggestions, feel free to contact us via email or asking a question through our [support center](https://ask.bioexcel.eu).\n",
        "\n",
        "And check also our [education web page](https://bonvinlab.org/education) where you will find more tutorials!\n"
      ],
      "metadata": {
        "id": "RVh9oEll_YRq"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "---\n",
        "\n",
        "#Save your results to Google Drive\n",
        "\n",
        "In order to save the entire tutorial directory with your results run the following cell."
      ],
      "metadata": {
        "id": "yB2L9u-YRwJW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "import os\n",
        "\n",
        "# Mount Google Drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "# Define the source directory\n",
        "source_dir = './HADDOCK3-antibody-antigen-notebook'\n",
        "\n",
        "# Define the destination directory in Google Drive\n",
        "destination_dir = '/content/drive/MyDrive/'\n",
        "\n",
        "# Create the destination directory if it doesn't exist\n",
        "os.makedirs(destination_dir, exist_ok=True)\n",
        "\n",
        "# Move the directory\n",
        "!cd ../; mv \"$source_dir\" \"$destination_dir\"\n",
        "\n",
        "print(f\"Moved '{source_dir}' to '{destination_dir}'\")"
      ],
      "metadata": {
        "id": "a2H4oaDuR_LO"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
